<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="GWASBrewer">
<title>Controlling Effect Size Distributions • GWASBrewer</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Controlling Effect Size Distributions">
<meta property="og:description" content="GWASBrewer">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">GWASBrewer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.0.0237</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/effect_distribution.html">Controlling Effect Size Distributions</a>
    <a class="dropdown-item" href="../articles/resampling.html">Resampling and Re-Scaling Summary and Individual Level Data</a>
    <a class="dropdown-item" href="../articles/simulating_data.html">Simulating Data</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/jean997/GWASBrewer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Controlling Effect Size Distributions</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jean997/GWASBrewer/blob/HEAD/vignettes/effect_distribution.Rmd" class="external-link"><code>vignettes/effect_distribution.Rmd</code></a></small>
      <div class="d-none name"><code>effect_distribution.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jean997/GWASBrewer" class="external-link">GWASBrewer</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hadley/reshape" class="external-link">reshape2</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><code>GWASBrewer</code> can support a range of effect size
distributions. At one end of the spectrum, effect sizes and direct
effect variants can be completely deterministic. At the other end of the
spectrum, both are random. There are multiple ways to control the
distributions of these random values demonstrated in this vignette.</p>
</div>
<div class="section level2">
<h2 id="default-behavior">Default Behavior<a class="anchor" aria-label="anchor" href="#default-behavior"></a>
</h2>
<p>The default behavior of <code>sim_mv</code> is to sample effect
variants randomly by first randomly selecting direct effect variants and
then choosing the standardized effect size of these variants from a
normal distribution with mean zero appropriately chosen variance.
Standardized effects are effects with units of change in outcome on SDs
per unit increase in genotype SD. This procedure results in effect
variants spread evenly throughout the genome and per-allele effect sizes
that are inversely proportional to <span class="math inline">\(\sqrt{2
f_j (1-f_j)}\)</span>.</p>
<p>We demonstrate this by simulating data for only one trait. We will
set <span class="math inline">\(N = 0\)</span> since we don’t need the
summary statistics for this demonstration.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">sim_dat1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_mv.html">sim_mv</a></span><span class="op">(</span>G <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                   J <span class="op">=</span> <span class="fl">50000</span>,</span>
<span>                   N <span class="op">=</span> <span class="fl">0</span>,  </span>
<span>                   h2 <span class="op">=</span> <span class="fl">0.3</span>, </span>
<span>                   pi <span class="op">=</span> <span class="fl">1000</span><span class="op">/</span><span class="fl">50000</span>, </span>
<span>                   af <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/Beta.html" class="external-link">rbeta</a></span><span class="op">(</span><span class="va">n</span>, <span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 50000 SNPs and 1 traits.</span></span></code></pre></div>
<p>In this specification, we expect to have about 1000 effect variants.
Below, we plot the distribution of the standardized and per-allele
effects. Note that in this simulation,
<code>direct_SNP_effects_joint</code> and <code>beta_joint</code> are
the same because there is only one trait.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">sim_dat1</span><span class="op">$</span><span class="va">direct_SNP_effects_joint</span> <span class="op">!=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1004</span></span>
<span></span>
<span><span class="va">ix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">sim_dat1</span><span class="op">$</span><span class="va">direct_SNP_effects_joint</span> <span class="op">!=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">per_allele_effects</span> <span class="op">&lt;-</span> <span class="va">sim_dat1</span><span class="op">$</span><span class="va">direct_SNP_effects_joint</span><span class="op">[</span><span class="va">ix</span><span class="op">]</span></span>
<span><span class="va">effect_afs</span> <span class="op">&lt;-</span> <span class="va">sim_dat1</span><span class="op">$</span><span class="va">snp_info</span><span class="op">$</span><span class="va">AF</span><span class="op">[</span><span class="va">ix</span><span class="op">]</span></span>
<span><span class="va">std_effects</span> <span class="op">&lt;-</span> <span class="va">per_allele_effects</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">2</span><span class="op">*</span><span class="va">effect_afs</span><span class="op">*</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">effect_afs</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## first plot per-allele effects</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">per_allele_effects</span>, breaks <span class="op">=</span> <span class="fl">30</span>, freq <span class="op">=</span> <span class="cn">F</span>, </span>
<span>     main <span class="op">=</span> <span class="st">"Per-Allele Direct Effects"</span><span class="op">)</span></span></code></pre></div>
<p><img src="effect_distribution_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## next show that standardized effects are normal</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">std_effects</span>, breaks <span class="op">=</span> <span class="fl">30</span>, freq <span class="op">=</span> <span class="cn">F</span>, </span>
<span>     main <span class="op">=</span> <span class="st">"Standardized Direct Effects"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/curve.html" class="external-link">curve</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">dnorm</a></span><span class="op">(</span><span class="va">x</span>, sd <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">0.3</span><span class="op">/</span><span class="fl">1000</span><span class="op">)</span><span class="op">)</span>, add <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="effect_distribution_files/figure-html/unnamed-chunk-3-2.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="controlling-which-variants-are-effect-variants">Controlling which Variants are Effect Variants<a class="anchor" aria-label="anchor" href="#controlling-which-variants-are-effect-variants"></a>
</h2>
<p>The <code>pi</code> argument to <code>sim_mv</code> gives the
probability that each variant is an effect variant. This can be a scalar
in which calse the probability is the same for all variants and all
traits, a vector giving a different probability for each trait or <span class="math inline">\(J \times M\)</span> matrix. If <code>pi</code> has
matrix format then <code>pi[j,m]</code> gives the probability than
variant <code>j</code> has a direct effect on trait <code>m</code>. To
make the set of direct effect variants fully deterministic, set the
values of <code>pi</code> to be either 0 or 1. Using <code>pi</code> in
matrix format is incompatible with options <code>pi_exact = TRUE</code>
and <code>sporadic_pleiotropy = FALSE</code>.</p>
<p>Below, we use the matrix format to force all effect variants to be in
the first half of the genome. This strategy could be used to designate
regions with high concentrations of effect variants or assign different
probabilities to variants with different features.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pi_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>, nrow <span class="op">=</span> <span class="fl">50000</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">pi_mat</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">25000</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1000</span><span class="op">/</span><span class="fl">25000</span></span>
<span><span class="va">sim_dat2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_mv.html">sim_mv</a></span><span class="op">(</span>G <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                   J <span class="op">=</span> <span class="fl">50000</span>,</span>
<span>                   N <span class="op">=</span> <span class="fl">60000</span>,  </span>
<span>                   h2 <span class="op">=</span> <span class="fl">0.3</span>, </span>
<span>                   pi <span class="op">=</span> <span class="va">pi_mat</span>, </span>
<span>                   af <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/Beta.html" class="external-link">rbeta</a></span><span class="op">(</span><span class="va">n</span>, <span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 50000 SNPs and 1 traits.</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sim_dat2</span><span class="op">$</span><span class="va">direct_SNP_effects_joint</span><span class="op">)</span></span></code></pre></div>
<p><img src="effect_distribution_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>In another application of this option, we can make the probability
that a variant is an effect variant a function of the allele
frequency.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">af</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Beta.html" class="external-link">rbeta</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">50000</span>, <span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">pi_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>, nrow <span class="op">=</span> <span class="fl">50000</span>, ncol <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">pi_mat</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmin</a></span><span class="op">(</span><span class="va">af</span>, <span class="fl">1</span><span class="op">-</span><span class="va">af</span><span class="op">)</span><span class="op">)</span><span class="op">^</span><span class="fl">4</span></span>
<span><span class="co"># ensure there are still 1000 effect variants expected</span></span>
<span><span class="va">pi_mat</span> <span class="op">&lt;-</span> <span class="fl">1000</span><span class="op">*</span><span class="va">pi_mat</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">pi_mat</span><span class="op">)</span> </span>
<span><span class="va">sim_dat3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_mv.html">sim_mv</a></span><span class="op">(</span>G <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                   J <span class="op">=</span> <span class="fl">50000</span>,</span>
<span>                   N <span class="op">=</span> <span class="fl">60000</span>,  </span>
<span>                   h2 <span class="op">=</span> <span class="fl">0.3</span>, </span>
<span>                   pi <span class="op">=</span> <span class="va">pi_mat</span>, </span>
<span>                   af <span class="op">=</span> <span class="va">af</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 50000 SNPs and 1 traits.</span></span></code></pre></div>
<p>To verify that we observe the desired effect, we plot the proportion
of effect variants by bins of minor allele frequency.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.5</span>, length.out <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmin</a></span><span class="op">(</span><span class="va">af</span>, <span class="fl">1</span><span class="op">-</span><span class="va">af</span><span class="op">)</span></span>
<span><span class="va">snp_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>maf <span class="op">=</span> <span class="va">maf</span>, </span>
<span>                      is_effect <span class="op">=</span> <span class="va">sim_dat3</span><span class="op">$</span><span class="va">direct_SNP_effects_joint</span> <span class="op">!=</span><span class="fl">0</span>, </span>
<span>                      maf_cat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cut.html" class="external-link">cut</a></span><span class="op">(</span><span class="va">maf</span>, breaks <span class="op">=</span> <span class="va">x</span><span class="op">)</span>, </span>
<span>                      maf_cat_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cut.html" class="external-link">cut</a></span><span class="op">(</span><span class="va">maf</span>, breaks <span class="op">=</span> <span class="va">x</span>, labels <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>           <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">maf_cat</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>           <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>, </span>
<span>                     n_eff <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">is_effect</span><span class="op">)</span>, </span>
<span>                     maf_min <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/nth.html" class="external-link">first</a></span><span class="op">(</span><span class="va">maf_cat_order</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>           <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>eff_prob <span class="op">=</span> <span class="va">n_eff</span><span class="op">/</span><span class="va">n</span><span class="op">)</span></span>
<span><span class="va">snp_dat</span><span class="op">$</span><span class="va">maf_min</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">[</span><span class="va">snp_dat</span><span class="op">$</span><span class="va">maf_min</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">snp_dat</span>, <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">maf_min</span>, <span class="va">eff_prob</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="effect_distribution_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="controlling-effect-size-distribution">Controlling Effect Size Distribution<a class="anchor" aria-label="anchor" href="#controlling-effect-size-distribution"></a>
</h2>
<p>To sample standardized effects from a distribution other than the
normal distribution, we use the <code>snp_effect_function</code>
argument in <code>sim_mv</code>. This argument can accept a function for
randomly sampling from any scale family distribution. The function given
to <code>snp_effect_function</code> must accept three arguments,
<code>n</code>, <code>sd</code>, and <code>snp_info</code>. The function
will be called internally by <code>sim_mv</code>. It should return a
length <code>n</code> vector of real numbers such that the expected sum
of squares is <code>sd^2</code>. That is, if <code>f</code> is our
function, and <code>x &lt;- f(n, sd, snp_info)</code> then
<code>x</code> should be length <code>n</code> and the expected value of
<code>sum(x^2)</code> should be <code>sd^2</code>.</p>
<p>The <code>snp_info</code> argument allows the function to receive
variant information such as allele frequency. The simplest effect size
functions will not use this argument at all, though all functions should
be able to accept it.</p>
<p>Currently, <code>GWASBrewer</code> will test that your function
executes and will test that it gives output with approximately the
expected variance for one test case. This warning will flag some major
errors but lack of warning does not guarantee that your function is
working correctly. You should be sure to verify it’s behavior for
yourself. You may see the warning occasionally for correctly working
functions due to randomness or if you function produces effect sizes
with an eccentric distribution.</p>
<div class="section level3">
<h3 id="drawing-effects-from-a-mixture-of-normals">Drawing Effects from a Mixture of Normals<a class="anchor" aria-label="anchor" href="#drawing-effects-from-a-mixture-of-normals"></a>
</h3>
<p>Frequently, we observe a pattern that many effect variants have a
small “background” effect and some have a larger “foreground” effect.
This pattern is consistent with an omnigenic model for genetic
architecture. To simulate this, we can draw standardized effects from a
mixture of normal distributions rather than from a single normal
distribution. There is a helper function
<code>mixnorm_to_scale_fam</code> which generates functions appropriate
to pass to <code>sim_mv</code>. For example, suppose we want a
distribution in which 90% of variants have standardized direct effects
following a <span class="math inline">\(N(0, \sigma^2_1)\)</span>
distribution and 10% have standardized effects from a <span class="math inline">\(N(0, 10 \sigma^2_1)\)</span> distribution. Then we
can use</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myf</span> <span class="op">&lt;-</span> <span class="fu">mixnorm_to_scale_fam</span><span class="op">(</span>sigma <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span>, pi <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.9</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Note that the <code>sigma</code> argument just specifies the relative
scaling of the variances so it would be equivalent to use</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myf</span> <span class="op">&lt;-</span> <span class="fu">mixnorm_to_scale_fam</span><span class="op">(</span>sigma <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">10</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span>, pi <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.9</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">sim_dat4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_mv.html">sim_mv</a></span><span class="op">(</span>G <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                   J <span class="op">=</span> <span class="fl">50000</span>,</span>
<span>                   N <span class="op">=</span> <span class="fl">0</span>,  </span>
<span>                   h2 <span class="op">=</span> <span class="fl">0.3</span>, </span>
<span>                   pi <span class="op">=</span> <span class="fl">5000</span><span class="op">/</span><span class="fl">50000</span>, </span>
<span>                   snp_effect_function <span class="op">=</span> <span class="va">myf</span>,</span>
<span>                   af <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/Beta.html" class="external-link">rbeta</a></span><span class="op">(</span><span class="va">n</span>, <span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 50000 SNPs and 1 traits.</span></span></code></pre></div>
<p>This gives a distribution of standardized direct effects with much
heavier tails than using the default behavior.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">std_effects</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">sim_dat4</span>, </span>
<span>                    <span class="va">direct_SNP_effects_joint</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">2</span><span class="op">*</span><span class="va">snp_info</span><span class="op">$</span><span class="va">AF</span><span class="op">*</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">snp_info</span><span class="op">$</span><span class="va">AF</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">std_effects</span><span class="op">[</span><span class="op">!</span><span class="va">std_effects</span> <span class="op">==</span> <span class="fl">0</span><span class="op">]</span>, breaks <span class="op">=</span> <span class="fl">30</span>, freq <span class="op">=</span> <span class="cn">F</span>, </span>
<span>     main <span class="op">=</span> <span class="st">"Standardized Direct Effects"</span><span class="op">)</span></span></code></pre></div>
<p><img src="effect_distribution_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>Rather than use the built-in helper function, we could have gotten
the same result using</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myf</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span>, <span class="va">sd</span>, <span class="va">...</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">Vtarget</span> <span class="op">&lt;-</span> <span class="va">sd</span><span class="op">^</span><span class="fl">2</span> <span class="co"># Target sum of squares</span></span>
<span>  <span class="va">Vbase</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fl">0.9</span><span class="op">*</span><span class="fl">1</span> <span class="op">+</span> <span class="fl">0.1</span><span class="op">*</span><span class="fl">10</span><span class="op">)</span><span class="op">*</span><span class="va">n</span> <span class="co"># base sum of squares</span></span>
<span>  <span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">Vtarget</span><span class="op">/</span><span class="va">Vbase</span><span class="op">)</span>  <span class="co"># scale factor</span></span>
<span>  <span class="va">mix_sds</span> <span class="op">&lt;-</span> <span class="va">a</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co"># choose which variants are from which component</span></span>
<span>  <span class="va">Z</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html" class="external-link">rbinom</a></span><span class="op">(</span>n <span class="op">=</span><span class="va">n</span>, size <span class="op">=</span> <span class="fl">1</span>, prob <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span>
<span>  <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="va">n</span><span class="op">)</span></span>
<span>  <span class="va">x</span><span class="op">[</span><span class="va">Z</span> <span class="op">==</span> <span class="fl">0</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">Z</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span>, sd <span class="op">=</span> <span class="va">mix_sds</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">x</span><span class="op">[</span><span class="va">Z</span> <span class="op">==</span> <span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">Z</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span>, sd <span class="op">=</span> <span class="va">mix_sds</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">## double check that we got the scaling right</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">myf</span><span class="op">(</span><span class="fl">1e5</span>, sd <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">0.2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span> <span class="co"># should be about 0.2</span></span></code></pre></div>
<p>Note that because our function has heavier tails, this particular
example will generate a warning about 2% of the time. Seeing this
warning doesn’t necessarily mean that anything is wrong. When in doubt,
check your code and check that the resulting heritability
(e.g. <code>sim_dat4$h2</code>) is close to the expected
heritability.</p>
</div>
<div class="section level3">
<h3 id="providing-a-fixed-list-of-relative-effect-sizes">Providing a fixed list of relative effect sizes<a class="anchor" aria-label="anchor" href="#providing-a-fixed-list-of-relative-effect-sizes"></a>
</h3>
<p>If you want to provide a specific fixed list of relative variant
effects, you can do this using the helper function
<code>fixed_to_scalefam</code>. Effects will still be scaled to give the
desired heritability, however the relative sizes of effects will be
given by a fixed list, repeated as many times as necessary to give the
desired number of variants. For example, the function below provides two
standardized effect sizes, one twice as big as the other. This is not a
very realistic distribution but it is ok for an example.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myf</span> <span class="op">&lt;-</span> <span class="fu">fixed_to_scale_fam</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">sim_dat5</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_mv.html">sim_mv</a></span><span class="op">(</span>G <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                   J <span class="op">=</span> <span class="fl">50000</span>,</span>
<span>                   N <span class="op">=</span> <span class="fl">0</span>,  </span>
<span>                   h2 <span class="op">=</span> <span class="fl">0.3</span>, </span>
<span>                   pi <span class="op">=</span> <span class="fl">5000</span><span class="op">/</span><span class="fl">50000</span>, </span>
<span>                   snp_effect_function <span class="op">=</span> <span class="va">myf</span>,</span>
<span>                   af <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/Beta.html" class="external-link">rbeta</a></span><span class="op">(</span><span class="va">n</span>, <span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 50000 SNPs and 1 traits.</span></span></code></pre></div>
<p>This gives us standardized effects that alternate between two
values.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">per_allele_effects</span> <span class="op">&lt;-</span> <span class="va">sim_dat5</span><span class="op">$</span><span class="va">direct_SNP_effects_joint</span></span>
<span><span class="va">genotype_variance</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">sim_dat5</span><span class="op">$</span><span class="va">snp_info</span>, <span class="fl">2</span><span class="op">*</span><span class="va">AF</span><span class="op">*</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">AF</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">std_effects</span> <span class="op">&lt;-</span> <span class="va">per_allele_effects</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">genotype_variance</span><span class="op">)</span></span>
<span><span class="va">std_effects</span><span class="op">[</span><span class="op">!</span><span class="va">std_effects</span> <span class="op">==</span> <span class="fl">0</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="co">#&gt;  [1] 0.004898979 0.009797959 0.004898979 0.009797959 0.004898979 0.009797959</span></span>
<span><span class="co">#&gt;  [7] 0.004898979 0.009797959 0.004898979 0.009797959</span></span></code></pre></div>
<p>Note that the per-allele effects take on many values because there is
variation in allele frequency.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">per_allele_effects</span><span class="op">[</span><span class="op">!</span><span class="va">per_allele_effects</span> <span class="op">==</span> <span class="fl">0</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span><span class="co">#&gt;  [1] 0.014672819 0.096632705 0.020872569 0.014834927 0.011288336 0.116741975</span></span>
<span><span class="co">#&gt;  [7] 0.009585994 0.013878979 0.008176130 0.020079876</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="providing-an-exact-set-of-direct-effects">Providing an Exact Set of Direct Effects<a class="anchor" aria-label="anchor" href="#providing-an-exact-set-of-direct-effects"></a>
</h3>
<p>The function <code>sim_mv_determined</code> is a wrapper function for
<code>sim_mv</code> that allows you to directly specify direct effects
for all traits and variants. The only differences between
<code>sim_mv_determined</code> and <code>sim_mv</code> are that
<code>sim_mv_determined</code> accepts the matrix of direct effects plus
parameters <code>geno_scale</code> and <code>pheno_sd</code> to indicate
the scale of the provided effects, and <code>sim_mv_determined</code>
does not accept parameters related to the distribution of effect sizes.
Note that for <code>sim_mv_determined</code>, the <code>G</code> matrix
gives effects on the phenotype scale of the effect sizes, indicated with
the <code>pheno_sd</code> argument. In <code>sim_mv</code>, the
phenotype scale is always 1 (phenotypes have unit variance). The
resulting object can be translated to a different phenotype scale using
<code>rescale_sumstats</code> if desired.</p>
<p>Below is a very simple example using <code>sim_mv_determined</code>
for ten variants and two traits. We will put the desired direct effects
into a matrix and then pass these to <code>sim_mv_determined</code>.
Here we use effects on the per-genotype SD scale, so we set
<code>geno_scale=sd</code>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">G</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.5</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span>, byrow <span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">my_effects</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>, nrow <span class="op">=</span> <span class="fl">10</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">my_effects</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.008</span>, <span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="va">my_effects</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">6</span>, <span class="fl">9</span><span class="op">)</span>, <span class="fl">2</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.02</span>, <span class="fl">0.06</span>, <span class="fl">0.009</span><span class="op">)</span></span>
<span><span class="va">my_effects</span></span>
<span><span class="co">#&gt;         [,1]   [,2]</span></span>
<span><span class="co">#&gt;  [1,] -0.008  0.000</span></span>
<span><span class="co">#&gt;  [2,]  0.000  0.000</span></span>
<span><span class="co">#&gt;  [3,]  0.000 -0.020</span></span>
<span><span class="co">#&gt;  [4,]  0.000  0.000</span></span>
<span><span class="co">#&gt;  [5,]  0.010  0.000</span></span>
<span><span class="co">#&gt;  [6,]  0.000  0.060</span></span>
<span><span class="co">#&gt;  [7,]  0.000  0.000</span></span>
<span><span class="co">#&gt;  [8,]  0.000  0.000</span></span>
<span><span class="co">#&gt;  [9,]  0.000  0.009</span></span>
<span><span class="co">#&gt; [10,]  0.000  0.000</span></span>
<span></span>
<span><span class="co"># for fun, lets include some sample overlap</span></span>
<span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">40000</span>, <span class="fl">10000</span>, <span class="fl">10000</span>, <span class="fl">20000</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">sim_dat6</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_mv_determined.html">sim_mv_determined</a></span><span class="op">(</span>N <span class="op">=</span> <span class="va">N</span>,</span>
<span>                              direct_SNP_effects_joint <span class="op">=</span> <span class="va">my_effects</span>,</span>
<span>                              geno_scale <span class="op">=</span> <span class="st">"sd"</span>,</span>
<span>                              pheno_sd <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                              G<span class="op">=</span><span class="va">G</span>,</span>
<span>                              est_s <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 10 SNPs and 2 traits.</span></span>
<span></span>
<span><span class="va">sim_dat6</span><span class="op">$</span><span class="va">direct_SNP_effects_joint</span></span>
<span><span class="co">#&gt;         [,1]   [,2]</span></span>
<span><span class="co">#&gt;  [1,] -0.008  0.000</span></span>
<span><span class="co">#&gt;  [2,]  0.000  0.000</span></span>
<span><span class="co">#&gt;  [3,]  0.000 -0.020</span></span>
<span><span class="co">#&gt;  [4,]  0.000  0.000</span></span>
<span><span class="co">#&gt;  [5,]  0.010  0.000</span></span>
<span><span class="co">#&gt;  [6,]  0.000  0.060</span></span>
<span><span class="co">#&gt;  [7,]  0.000  0.000</span></span>
<span><span class="co">#&gt;  [8,]  0.000  0.000</span></span>
<span><span class="co">#&gt;  [9,]  0.000  0.009</span></span>
<span><span class="co">#&gt; [10,]  0.000  0.000</span></span>
<span><span class="va">sim_dat6</span><span class="op">$</span><span class="va">beta_joint</span></span>
<span><span class="co">#&gt;         [,1]   [,2]</span></span>
<span><span class="co">#&gt;  [1,] -0.008 -0.004</span></span>
<span><span class="co">#&gt;  [2,]  0.000  0.000</span></span>
<span><span class="co">#&gt;  [3,]  0.000 -0.020</span></span>
<span><span class="co">#&gt;  [4,]  0.000  0.000</span></span>
<span><span class="co">#&gt;  [5,]  0.010  0.005</span></span>
<span><span class="co">#&gt;  [6,]  0.000  0.060</span></span>
<span><span class="co">#&gt;  [7,]  0.000  0.000</span></span>
<span><span class="co">#&gt;  [8,]  0.000  0.000</span></span>
<span><span class="co">#&gt;  [9,]  0.000  0.009</span></span>
<span><span class="co">#&gt; [10,]  0.000  0.000</span></span>
<span><span class="va">sim_dat6</span><span class="op">$</span><span class="va">Sigma_G</span></span>
<span><span class="co">#&gt;          [,1]     [,2]</span></span>
<span><span class="co">#&gt; [1,] 0.000164 0.000082</span></span>
<span><span class="co">#&gt; [2,] 0.000082 0.004122</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="different-effect-distributions-for-different-traits">Different effect distributions for different traits<a class="anchor" aria-label="anchor" href="#different-effect-distributions-for-different-traits"></a>
</h3>
<p>The <code>snp_effect_function</code> parameter can accept a list with
length equal to the number of traits. In this case, the direct effect
variants for each trait will be drawn from the corresponding element of
<code>snp_effect_function</code>. If <code>snp_effect_function</code> is
a list, each element should either be an acceptable function or the
string <code>normal</code> to use the default option. The code below
generates data for two independent traits. The effects for the first
trait are normally distributed. The effects for the second trait take on
one of two values as in the previous section.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myf</span> <span class="op">&lt;-</span> <span class="fu">fixed_to_scale_fam</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sim_dat7</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_mv.html">sim_mv</a></span><span class="op">(</span>G <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                   J <span class="op">=</span> <span class="fl">50000</span>,</span>
<span>                   N <span class="op">=</span> <span class="fl">60000</span>,  </span>
<span>                   h2 <span class="op">=</span> <span class="fl">0.3</span>, </span>
<span>                   pi <span class="op">=</span> <span class="fl">5000</span><span class="op">/</span><span class="fl">50000</span>, </span>
<span>                   snp_effect_function <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'normal'</span>, <span class="va">myf</span><span class="op">)</span>,</span>
<span>                   af <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/Beta.html" class="external-link">rbeta</a></span><span class="op">(</span><span class="va">n</span>, <span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 50000 SNPs and 2 traits.</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">std_effects1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">sim_dat7</span>, </span>
<span>                     <span class="va">direct_SNP_effects_joint</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">2</span><span class="op">*</span><span class="va">snp_info</span><span class="op">$</span><span class="va">AF</span><span class="op">*</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">snp_info</span><span class="op">$</span><span class="va">AF</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">std_effects1</span><span class="op">[</span><span class="op">!</span><span class="va">std_effects1</span> <span class="op">==</span> <span class="fl">0</span><span class="op">]</span>, breaks <span class="op">=</span> <span class="fl">30</span>, freq <span class="op">=</span> <span class="cn">F</span>, main <span class="op">=</span> <span class="st">"Standardized Direct Effects Trait 1"</span><span class="op">)</span></span></code></pre></div>
<p><img src="effect_distribution_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">std_effects2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">sim_dat7</span>, </span>
<span>                     <span class="va">direct_SNP_effects_joint</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">2</span><span class="op">*</span><span class="va">snp_info</span><span class="op">$</span><span class="va">AF</span><span class="op">*</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">snp_info</span><span class="op">$</span><span class="va">AF</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">std_effects2</span><span class="op">[</span><span class="op">!</span><span class="va">std_effects2</span> <span class="op">==</span> <span class="fl">0</span><span class="op">]</span>, breaks <span class="op">=</span> <span class="fl">30</span>, freq <span class="op">=</span> <span class="cn">F</span>, main <span class="op">=</span> <span class="st">"Standardized Direct Effects Trait 2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="effect_distribution_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="example-using-pre-specified-effects">Example Using Pre-Specified Effects<a class="anchor" aria-label="anchor" href="#example-using-pre-specified-effects"></a>
</h3>
<p>Below is some code that combines some of the effect size
specification options we have seen to simulate GWAS data for one LD
block and two traits. The first trait will have two effect variants that
have equal standardized effect size. The second trait has two effect
variants with different effect sizes. One pair of variants co-localizes
across traits. Simulations of this type could be used to test
co-localization methods. Since we are looking at only one block, we will
set the heritability of the block to be a reasonable amount of
heritability provided by two variants, 6e-4 for one trait and 1e-3 for
the other.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ld_mat_list"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"AF"</span><span class="op">)</span></span>
<span><span class="co"># Pull out sub section of one of the built in LD blocks</span></span>
<span><span class="va">ld_block</span> <span class="op">&lt;-</span> <span class="va">ld_mat_list</span><span class="op">[[</span><span class="fl">10</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">118</span><span class="op">:</span><span class="fl">350</span>, <span class="fl">118</span><span class="op">:</span><span class="fl">350</span><span class="op">]</span></span>
<span><span class="va">J</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">ld_block</span><span class="op">)</span></span>
<span><span class="co"># Corresponding allele frequency</span></span>
<span><span class="va">start_ix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">ld_mat_list</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span>, <span class="va">nrow</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fl">118</span> </span>
<span><span class="va">af</span> <span class="op">&lt;-</span> <span class="va">AF</span><span class="op">[</span><span class="va">start_ix</span><span class="op">:</span><span class="op">(</span><span class="va">start_ix</span> <span class="op">+</span> <span class="va">J</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">## plot the LD block</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">ld_block</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu">reshape2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html" class="external-link">melt</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Var1</span>, y <span class="op">=</span> <span class="va">Var2</span>, fill <span class="op">=</span> <span class="va">value</span><span class="op">)</span><span class="op">)</span>  <span class="op">+</span> </span>
<span>      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_tile</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_fill_gradient2</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="effect_distribution_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Select causal variants</span></span>
<span><span class="va">pi_mat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>, nrow <span class="op">=</span> <span class="va">J</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">pi_mat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">75</span>, <span class="fl">150</span><span class="op">)</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="va">pi_mat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">75</span>, <span class="fl">130</span><span class="op">)</span>, <span class="fl">2</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span></span>
<span><span class="co"># Set effect size functions</span></span>
<span><span class="va">myf1</span> <span class="op">&lt;-</span> <span class="fu">fixed_to_scale_fam</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">myf2</span> <span class="op">&lt;-</span> <span class="fu">fixed_to_scale_fam</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sim_dat7</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_mv.html">sim_mv</a></span><span class="op">(</span>G <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                   J <span class="op">=</span> <span class="va">J</span>,</span>
<span>                   N <span class="op">=</span> <span class="fl">100000</span>,  </span>
<span>                   h2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">6e-4</span>, <span class="fl">1e-3</span><span class="op">)</span>, </span>
<span>                   pi <span class="op">=</span> <span class="va">pi_mat</span>,</span>
<span>                   R_LD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">ld_block</span><span class="op">)</span>,</span>
<span>                   snp_effect_function <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">myf1</span>, <span class="va">myf2</span><span class="op">)</span>,</span>
<span>                   af <span class="op">=</span> <span class="va">af</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 233 SNPs and 2 traits.</span></span>
<span></span>
<span><span class="co">## Plotting joint effects</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sim_dat7</span><span class="op">$</span><span class="va">direct_SNP_effects_joint</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, main <span class="op">=</span> <span class="st">"Trait 1 Joint Effects"</span>, ylab <span class="op">=</span> <span class="st">"beta joint"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>v <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">75</span>, <span class="fl">150</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="effect_distribution_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sim_dat7</span><span class="op">$</span><span class="va">direct_SNP_effects_joint</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, main <span class="op">=</span> <span class="st">"Trait 2 Joint Effects"</span>, ylab <span class="op">=</span> <span class="st">"beta joint"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>v <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">75</span>, <span class="fl">130</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="effect_distribution_files/figure-html/unnamed-chunk-22-2.png" width="700"></p>
<p>Note that the effects for the first trait are not the same because we
specified standardized effects but have plotted per-allele effects.
Below we plot the marginal associations and observed <span class="math inline">\(z\)</span>-scores for this data.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plotting marginal effects</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sim_dat7</span><span class="op">$</span><span class="va">direct_SNP_effects_marg</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, main <span class="op">=</span> <span class="st">"Trait 1 Marginal Effects"</span>, ylab <span class="op">=</span> <span class="st">"beta marg"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>v <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">75</span>, <span class="fl">150</span><span class="op">)</span>, h <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="effect_distribution_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sim_dat7</span><span class="op">$</span><span class="va">direct_SNP_effects_marg</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, main <span class="op">=</span> <span class="st">"Trait 2 Marginal Effects"</span>, ylab <span class="op">=</span> <span class="st">"beta marg"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>v <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">75</span>, <span class="fl">130</span><span class="op">)</span>, h <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="effect_distribution_files/figure-html/unnamed-chunk-23-2.png" width="700"></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plotting observed z-scores</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">sim_dat7</span>, <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">beta_hat</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">/</span><span class="va">se_beta_hat</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, main <span class="op">=</span> <span class="st">"Trait 1 Obs. Z-scores"</span>, ylab <span class="op">=</span> <span class="st">"zhat marg"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>v <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">75</span>, <span class="fl">150</span><span class="op">)</span>, h <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="effect_distribution_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">sim_dat7</span>, <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">beta_hat</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">/</span><span class="va">se_beta_hat</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, main <span class="op">=</span> <span class="st">"Trait 2 Obs. Z-scores"</span>, ylab <span class="op">=</span> <span class="st">"zhat marg"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>v <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">75</span>, <span class="fl">130</span><span class="op">)</span>, h <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="effect_distribution_files/figure-html/unnamed-chunk-24-2.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="custom-effect-size-distributions">Custom Effect Size Distributions<a class="anchor" aria-label="anchor" href="#custom-effect-size-distributions"></a>
</h3>
<p>If you write your own function to pass to
<code>snp_effect_function</code>, it needs to accept three parameters:
number of variants (<code>n</code>), standard deviation (square root of
total variance explained) (<code>sd</code>), and SNP information
(<code>snp_info</code>). This last argument can be used to create effect
distributions that depend on allele frequency or other features.</p>
<p>The default model used by <code>GWASBrewer</code> is a sparse version
of the GCTA model in which all causal variants have the same expected
heritability explained. We might, instead want to use a heritability
model like the <a href="https://dougspeed.com/" class="external-link">LDAK</a> model or we
might want to allow the expected heritability explained to depend on an
annotation.</p>
<p>For this example, we assume that we would like to use the
heritability model</p>
<p><span class="math display">\[E[h_j^2] = x(f_j (1-f_j))^{0.75}(0.1
A_{1,j} + 0.13 A_{2,j} + 0.9)\]</span></p>
<p>where <span class="math inline">\(x\)</span> is a scalar that depends
on the total heritability and <span class="math inline">\(A_{1,j}\)</span> and <span class="math inline">\(A_{2,j}\)</span> are the values of two annotations
for variant <span class="math inline">\(j\)</span>. This is a simplified
version of the LDAK model. To use this model, we need a data frame
containing the annotations that we will pass to the <code>sim_mv</code>
function via the <code>snp_info</code> argument. If you are using LD,
the input <code>snp_info</code> data frame should have number of rows
equal to the size of the LD pattern. Otherwise, it should have number of
rows equal to the number of variants. For this example, we will generate
the two annotations randomly and will not use LD. To really simulate
from the LDAK model, it would be better to generate data with LD and
realistic annotations.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_snp_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>A1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Exponential.html" class="external-link">rexp</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">50000</span>, rate <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, </span>
<span>                          A2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Exponential.html" class="external-link">rexp</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">50000</span>, rate <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Next, we write a function to generate effect sizes. The
<code>snp_info</code> data frame that our function receives will include
the annotations given to the input <code>snp_info</code> argument of
<code>sim_mv</code> <strong>plus</strong> columns called
<code>SNP</code> for SNP name (usually not useful) and <code>AF</code>
for allele frequency.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">myf</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span>, <span class="va">sd</span>, <span class="va">snp_info</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">snp_info</span><span class="op">)</span> <span class="op">==</span> <span class="va">n</span><span class="op">)</span></span>
<span>  <span class="co"># expected heritability explained by each variant up to a scalar</span></span>
<span>  <span class="co"># AF, A1, and A2 are all columns of snp_info</span></span>
<span>  <span class="co"># we provided A1 and A2. sim_mv added the AF column.</span></span>
<span>  <span class="va">s2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">snp_info</span>, </span>
<span>             <span class="op">(</span><span class="op">(</span><span class="va">AF</span><span class="op">*</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">AF</span><span class="op">)</span><span class="op">)</span><span class="op">^</span><span class="fl">0.75</span><span class="op">)</span><span class="op">*</span><span class="op">(</span><span class="fl">0.1</span><span class="op">*</span><span class="va">A1</span> <span class="op">+</span> <span class="fl">0.13</span><span class="op">*</span><span class="va">A2</span> <span class="op">+</span> <span class="fl">0.9</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co"># Scale s2 to give sum(s2) = sd^2</span></span>
<span>  <span class="va">scale</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">sd</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">s2</span><span class="op">)</span></span>
<span>  <span class="va">s2</span> <span class="op">&lt;-</span> <span class="va">s2</span><span class="op">*</span><span class="va">scale</span></span>
<span>  <span class="co"># sample standardized effect sizes from normal distributions</span></span>
<span>  <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">n</span>, mean <span class="op">=</span> <span class="fl">0</span>, sd <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">s2</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Now we generate data,</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">sim_dat8</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_mv.html">sim_mv</a></span><span class="op">(</span>G <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                    J <span class="op">=</span> <span class="fl">50000</span>, </span>
<span>                    N <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                    h2 <span class="op">=</span> <span class="fl">0.3</span>, </span>
<span>                    pi <span class="op">=</span> <span class="fl">1</span>, </span>
<span>                    snp_effect_function <span class="op">=</span> <span class="va">myf</span>,</span>
<span>                    af <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/Beta.html" class="external-link">rbeta</a></span><span class="op">(</span><span class="va">n</span>, <span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span><span class="op">}</span>,</span>
<span>                    snp_info <span class="op">=</span> <span class="va">my_snp_info</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 50000 SNPs and 1 traits.</span></span></code></pre></div>
<p>The annotations will be included in the <code>snp_info</code> data
frame included in the output along with allele frequency and SNP
number.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sim_dat8</span><span class="op">$</span><span class="va">snp_info</span><span class="op">)</span></span>
<span><span class="co">#&gt;          A1         A2 SNP         AF</span></span>
<span><span class="co">#&gt; 1 0.7146837 0.09386619   1 0.16255313</span></span>
<span><span class="co">#&gt; 2 1.7789936 0.40783678   2 0.21165824</span></span>
<span><span class="co">#&gt; 3 0.6221027 0.88799175   3 0.34577086</span></span>
<span><span class="co">#&gt; 4 0.1222602 0.41385261   4 0.11092564</span></span>
<span><span class="co">#&gt; 5 1.9965800 0.37079998   5 0.09658463</span></span>
<span><span class="co">#&gt; 6 0.8135747 1.53049069   6 0.26393802</span></span></code></pre></div>
<p>Let’s verify that our function did what we thought it would by
plotting the average heritability explained by bins of <span class="math inline">\((f_j (1-f_j))^{0.75}(0.1 A_{1,j} + 0.13 A_{2,j} +
0.9)\)</span>.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_snp_info</span> <span class="op">&lt;-</span> <span class="va">sim_dat8</span><span class="op">$</span><span class="va">snp_info</span></span>
<span><span class="va">my_snp_info</span><span class="op">$</span><span class="va">beta_joint</span> <span class="op">&lt;-</span> <span class="va">sim_dat8</span><span class="op">$</span><span class="va">beta_joint</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">my_snp_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="va">my_snp_info</span>, </span>
<span>                      exp_h2_factor <span class="op">=</span> <span class="op">(</span><span class="op">(</span><span class="va">AF</span><span class="op">*</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">AF</span><span class="op">)</span><span class="op">)</span><span class="op">^</span><span class="fl">0.75</span><span class="op">)</span> <span class="op">*</span><span class="op">(</span> <span class="fl">0.1</span><span class="op">*</span><span class="va">A1</span> <span class="op">+</span> <span class="fl">0.13</span><span class="op">*</span><span class="va">A2</span> <span class="op">+</span> <span class="fl">0.9</span><span class="op">)</span>,</span>
<span>                      h2_explained <span class="op">=</span> <span class="fl">2</span><span class="op">*</span><span class="va">AF</span><span class="op">*</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">AF</span><span class="op">)</span><span class="op">*</span><span class="op">(</span><span class="va">beta_joint</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">my_snp_info</span><span class="op">$</span><span class="va">exp_h2_bin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cut.html" class="external-link">cut</a></span><span class="op">(</span><span class="va">my_snp_info</span><span class="op">$</span><span class="va">exp_h2_factor</span>, breaks <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span>
<span><span class="va">my_snp_info</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>       <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">exp_h2_bin</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>       <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>mean_h2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">h2_explained</span><span class="op">)</span>,</span>
<span>                 mean_exp_h2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">exp_h2_factor</span><span class="op">)</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">mean_exp_h2</span>, y <span class="op">=</span> <span class="va">mean_h2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Average Expected h2 (up to scale factor)"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Average observed h2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="effect_distribution_files/figure-html/unnamed-chunk-29-1.png" width="700"></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jean Morrison.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
