<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="GWASBrewer">
<title>Resampling and Re-Scaling Summary and Individual Level Data • GWASBrewer</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Resampling and Re-Scaling Summary and Individual Level Data">
<meta property="og:description" content="GWASBrewer">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">GWASBrewer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.0.0237</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/effect_distribution.html">Controlling Effect Size Distributions</a>
    <a class="dropdown-item" href="../articles/resampling.html">Resampling and Re-Scaling Summary and Individual Level Data</a>
    <a class="dropdown-item" href="../articles/simulating_data.html">Simulating Data</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/jean997/GWASBrewer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Resampling and Re-Scaling Summary and Individual Level Data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jean997/GWASBrewer/blob/HEAD/vignettes/resampling.Rmd" class="external-link"><code>vignettes/resampling.Rmd</code></a></small>
      <div class="d-none name"><code>resampling.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jean997/GWASBrewer" class="external-link">GWASBrewer</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In some applications, we may need to generate multiple data sets for
the same set of effect sizes. This would mimic performing multiple GWAS
on the same trait. There are two functions,
<code>resample_sumstats</code> and <code>resample_inddata</code> that
facilitate this. This vignette will demonstrate these functions as well
as discuss important considerations when the simulation involves
sampling GWAS for multiple ancestries for the same trait.</p>
<p>To get started, we will generate a small <code>sim_mv</code> object.
For a review of that function, see the “Simulating Data” vignette. For
this vignette, we will keep things very simple by having only 12
variants with very large effect sizes. Of course, this is not realistic,
but it will let us see exactly what is going on. In our example, we will
have two traits. The first trait has an effect of 0.2 on the second
trait. Recalling that <code>sim_mv</code> always sets the variance of
each trait to 1, this means, that trait 1 explains 4% (<span class="math inline">\(0.2^2 = 0.04\)</span>) of the variance of trait 2.
We will also use an LD pattern. Here we just specify the LD pattern for
five variants. This will be repeated 2.2 times to cover 12 variants. We
will specify <code>N</code> so that there is no sample overlap.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">my_ld1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span> <span class="fl">1.00</span>,  <span class="fl">0.60</span>,  <span class="fl">0.40</span>, <span class="op">-</span><span class="fl">0.1</span>, <span class="op">-</span><span class="fl">0.07</span>,</span>
<span>                 <span class="fl">0.60</span>,  <span class="fl">1.00</span>,  <span class="fl">0.60</span>, <span class="op">-</span><span class="fl">0.1</span>, <span class="op">-</span><span class="fl">0.07</span>,</span>
<span>                 <span class="fl">0.40</span>,  <span class="fl">0.60</span>,  <span class="fl">1.00</span>, <span class="op">-</span><span class="fl">0.1</span>, <span class="op">-</span><span class="fl">0.07</span>,</span>
<span>                 <span class="op">-</span><span class="fl">0.10</span>, <span class="op">-</span><span class="fl">0.10</span>, <span class="op">-</span><span class="fl">0.10</span>,  <span class="fl">1.0</span>,  <span class="fl">0.90</span>,</span>
<span>                 <span class="op">-</span><span class="fl">0.07</span>, <span class="op">-</span><span class="fl">0.07</span>, <span class="op">-</span><span class="fl">0.07</span>,  <span class="fl">0.9</span>,  <span class="fl">1.00</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">5</span>, byrow <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">af1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.35</span>, <span class="fl">0.3</span>, <span class="fl">0.4</span>, <span class="fl">0.72</span>, <span class="fl">0.75</span><span class="op">)</span></span>
<span></span>
<span><span class="va">G</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.2</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span>, byrow <span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="co"># matrix of causal effects</span></span>
<span><span class="va">orig_dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_mv.html">sim_mv</a></span><span class="op">(</span>N <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10000</span>, <span class="fl">20000</span><span class="op">)</span>,</span>
<span>                 J <span class="op">=</span> <span class="fl">12</span>,</span>
<span>                 h2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="fl">0.3</span><span class="op">)</span>,</span>
<span>                 pi <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span>                 G <span class="op">=</span> <span class="va">G</span>,</span>
<span>                 R_LD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">my_ld1</span><span class="op">)</span>,</span>
<span>                 af <span class="op">=</span> <span class="va">af1</span>, </span>
<span>                 est_s <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 12 SNPs and 2 traits.</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="resampling-summary-statistics-or-individual-level-data-from-the-same-population">Resampling Summary Statistics or Individual Level Data from the Same
Population<a class="anchor" aria-label="anchor" href="#resampling-summary-statistics-or-individual-level-data-from-the-same-population"></a>
</h2>
<p>In the simplest scenario, we want to simulate performing a new GWAS
or collecting new individual level data for the original set of traits
in the original population, meaning that the LD pattern and allele
frequencies are the same.</p>
<div class="section level3">
<h3 id="resampling-summary-statistics-from-the-same-population">Resampling Summary Statistics from the Same Population<a class="anchor" aria-label="anchor" href="#resampling-summary-statistics-from-the-same-population"></a>
</h3>
<p>To resample summary statistics, we can use
<code>resample_sumstats</code>. Here, we will regenerate data assuming
that samples for the new GWAS are totally overlapping.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">N1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">50000</span>, nrow <span class="op">=</span> <span class="fl">2</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">new_dat1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/resample_sumstats.html">resample_sumstats</a></span><span class="op">(</span>dat <span class="op">=</span> <span class="va">orig_dat</span>,</span>
<span>                              N <span class="op">=</span> <span class="va">N1</span>, </span>
<span>                              R_LD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">my_ld1</span><span class="op">)</span>, </span>
<span>                              af <span class="op">=</span> <span class="va">af1</span>, </span>
<span>                              est_s <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; I will assume that the environmental variance is the same in the old and new population.</span></span>
<span><span class="co">#&gt; I will assume that environmental correlation is the same in the old and new population. Note that this could result in different overall trait correlations.</span></span>
<span><span class="co">#&gt; SNP effects provided for 12 SNPs and 2 traits.</span></span></code></pre></div>
<p>The <code>new_dat1</code> object is another object of class
<code>sim_mv</code>. The new and old data have the same effect sizes so
<code>beta_joint</code>, <code>beta_marg</code>,
<code>direct_SNP_effects_joint</code>, and
<code>direct_SNP_effects_marg</code> are all the same between the two
objects. They also have the same causal effect structure so the trait
effect matrices <code>direct_trait_effects</code> and
<code>total_trait_effects</code> are the same. However, the summary
statistic matrices <code>beta_hat</code> and <code>s_estimate</code> are
different due to random sampling variation. The true standard errors,
<code>se_beta_hat</code> are also different because the sample size is
different. Finally, since there is sample overlap in the samples for
<code>new_dat1</code> but not in <code>orig_dat</code>, the
<code>R</code> matrices which give the correlation of effect sizes
across traits are different.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html" class="external-link">all.equal</a></span><span class="op">(</span><span class="va">orig_dat</span><span class="op">$</span><span class="va">beta_joint</span>, <span class="va">new_dat1</span><span class="op">$</span><span class="va">beta_joint</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html" class="external-link">all.equal</a></span><span class="op">(</span><span class="va">orig_dat</span><span class="op">$</span><span class="va">beta_marg</span>, <span class="va">new_dat1</span><span class="op">$</span><span class="va">beta_marg</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">orig_dat</span><span class="op">$</span><span class="va">beta_hat</span><span class="op">)</span></span>
<span><span class="co">#&gt;             [,1]         [,2]</span></span>
<span><span class="co">#&gt; [1,] -0.11039241  0.004995358</span></span>
<span><span class="co">#&gt; [2,] -0.18785203  0.002766487</span></span>
<span><span class="co">#&gt; [3,] -0.11483385  0.018581987</span></span>
<span><span class="co">#&gt; [4,]  0.22595316  0.039173114</span></span>
<span><span class="co">#&gt; [5,]  0.19996902  0.037651726</span></span>
<span><span class="co">#&gt; [6,] -0.01467208 -0.304974486</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">new_dat1</span><span class="op">$</span><span class="va">beta_hat</span><span class="op">)</span></span>
<span><span class="co">#&gt;              [,1]          [,2]</span></span>
<span><span class="co">#&gt; [1,] -0.116203623  0.0092874402</span></span>
<span><span class="co">#&gt; [2,] -0.175604708  0.0220111130</span></span>
<span><span class="co">#&gt; [3,] -0.110532292 -0.0007815694</span></span>
<span><span class="co">#&gt; [4,]  0.251954283  0.0446050591</span></span>
<span><span class="co">#&gt; [5,]  0.231799766  0.0426176263</span></span>
<span><span class="co">#&gt; [6,]  0.005154759 -0.3105532234</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">orig_dat</span><span class="op">$</span><span class="va">se_beta_hat</span><span class="op">)</span></span>
<span><span class="co">#&gt;            [,1]       [,2]</span></span>
<span><span class="co">#&gt; [1,] 0.01482499 0.01048285</span></span>
<span><span class="co">#&gt; [2,] 0.01543033 0.01091089</span></span>
<span><span class="co">#&gt; [3,] 0.01443376 0.01020621</span></span>
<span><span class="co">#&gt; [4,] 0.01574852 0.01113589</span></span>
<span><span class="co">#&gt; [5,] 0.01632993 0.01154701</span></span>
<span><span class="co">#&gt; [6,] 0.01482499 0.01048285</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">new_dat1</span><span class="op">$</span><span class="va">se_beta_hat</span><span class="op">)</span></span>
<span><span class="co">#&gt;             [,1]        [,2]</span></span>
<span><span class="co">#&gt; [1,] 0.006629935 0.006629935</span></span>
<span><span class="co">#&gt; [2,] 0.006900656 0.006900656</span></span>
<span><span class="co">#&gt; [3,] 0.006454972 0.006454972</span></span>
<span><span class="co">#&gt; [4,] 0.007042952 0.007042952</span></span>
<span><span class="co">#&gt; [5,] 0.007302967 0.007302967</span></span>
<span><span class="co">#&gt; [6,] 0.006629935 0.006629935</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">orig_dat</span><span class="op">$</span><span class="va">s_estimate</span><span class="op">)</span></span>
<span><span class="co">#&gt;            [,1]       [,2]</span></span>
<span><span class="co">#&gt; [1,] 0.01487042 0.01046558</span></span>
<span><span class="co">#&gt; [2,] 0.01514898 0.01089173</span></span>
<span><span class="co">#&gt; [3,] 0.01437689 0.01018576</span></span>
<span><span class="co">#&gt; [4,] 0.01556285 0.01122621</span></span>
<span><span class="co">#&gt; [5,] 0.01620126 0.01160266</span></span>
<span><span class="co">#&gt; [6,] 0.01489534 0.01026832</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">new_dat1</span><span class="op">$</span><span class="va">s_estimate</span><span class="op">)</span></span>
<span><span class="co">#&gt;             [,1]        [,2]</span></span>
<span><span class="co">#&gt; [1,] 0.006602645 0.006619905</span></span>
<span><span class="co">#&gt; [2,] 0.006844339 0.006885392</span></span>
<span><span class="co">#&gt; [3,] 0.006414776 0.006430850</span></span>
<span><span class="co">#&gt; [4,] 0.006920035 0.007005126</span></span>
<span><span class="co">#&gt; [5,] 0.007188453 0.007256993</span></span>
<span><span class="co">#&gt; [6,] 0.006635957 0.006485925</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="resampling-individual-level-data-from-the-same-population">Resampling Individual Level Data from the Same Population<a class="anchor" aria-label="anchor" href="#resampling-individual-level-data-from-the-same-population"></a>
</h3>
<p>We can also generate individual level data. For this we use the
<code>resample_inddata</code> function. This function can be used to
produce three types of output:</p>
<ul>
<li>Genotypes only</li>
<li>Genotypes and phenotypes</li>
<li>Phenotypes only</li>
</ul>
<p><code>resample_inddata</code> uses the <code>hapsim</code> package to
generate individual level data. The sample size argument,
<code>N</code>, in <code>resample_inddata</code> can accept three types
of input: scalar, vector, or data frame. These formats are the same as
those used by <code>sim_mv</code>, so if <code>N</code> is a scalar or
vector, <code>resample_inddata</code> will assume there is no sample
overlap. This means that if we use <code>N = 10</code> with two traits,
we will get out genotype information for 20 individuals, of which half
have phenotype 1 information and half have phenotype 2 information. To
specify sample overlap, we can use the data frame format. As a reminder,
a sample size data frame should have columns named <code>trait_1</code>,
… <code>trait_[M]</code> and <code>N</code>. The <code>trait_[x]</code>
columns will be interpreted as logicals and the <code>N</code> column
should give the number of samples in each combination of studies.</p>
<div class="section level4">
<h4 id="generating-genotypes-only">Generating genotypes only<a class="anchor" aria-label="anchor" href="#generating-genotypes-only"></a>
</h4>
<p>If <code>resample_inddata</code> is used to generate genotypes only,
we are not really “resampling” anything because there is no need to
input trait data. In this case, we only need to supply the total number
of individuals, the number of variants, and the LD pattern if desired.
Here we will generate data for 15 individuals with an LD pattern
matching the one used in <code>orig_dat</code>. The returned data
includes the genotype matrix and a vector of population allele
frequencies.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genos_only</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/resample_inddata.html">resample_inddata</a></span><span class="op">(</span>N <span class="op">=</span> <span class="fl">15</span>, </span>
<span>                               J <span class="op">=</span> <span class="fl">12</span>,</span>
<span>                               R_LD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">my_ld1</span><span class="op">)</span>, </span>
<span>                               af <span class="op">=</span> <span class="va">af1</span><span class="op">)</span></span>
<span><span class="co">#&gt; Generating genotype matrix only.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">genos_only</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "X"  "af"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">genos_only</span><span class="op">$</span><span class="va">X</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 15 12</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">genos_only</span><span class="op">$</span><span class="va">af</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 12</span></span>
<span><span class="va">genos_only</span><span class="op">$</span><span class="va">X</span></span>
<span><span class="co">#&gt;     [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12]</span></span>
<span><span class="co">#&gt; S1     2    1    1    1    1    0    0    0    2     2     1     1</span></span>
<span><span class="co">#&gt; S2     1    0    0    2    2    1    1    0    1     1     2     2</span></span>
<span><span class="co">#&gt; S3     0    0    0    0    1    1    1    1    1     1     2     2</span></span>
<span><span class="co">#&gt; S4     0    0    0    1    1    0    0    0    0     0     2     2</span></span>
<span><span class="co">#&gt; S5     0    0    0    2    2    0    1    0    1     2     2     1</span></span>
<span><span class="co">#&gt; S6     1    1    1    0    0    0    0    1    1     1     1     1</span></span>
<span><span class="co">#&gt; S7     0    0    1    0    0    1    0    1    1     1     2     2</span></span>
<span><span class="co">#&gt; S8     1    1    1    2    2    1    0    1    2     2     0     2</span></span>
<span><span class="co">#&gt; S9     0    0    0    1    1    1    0    0    2     2     1     1</span></span>
<span><span class="co">#&gt; S10    1    1    1    2    2    1    1    1    2     2     1     1</span></span>
<span><span class="co">#&gt; S11    1    0    0    1    1    0    0    1    1     1     2     2</span></span>
<span><span class="co">#&gt; S12    1    2    2    1    1    1    1    2    2     2     2     2</span></span>
<span><span class="co">#&gt; S13    0    0    0    2    1    0    0    1    1     1     2     2</span></span>
<span><span class="co">#&gt; S14    0    0    0    2    2    1    1    1    2     2     1     1</span></span>
<span><span class="co">#&gt; S15    0    0    0    2    2    1    0    2    0     1     2     2</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="generating-genotypes-and-phenotypes">Generating Genotypes and Phenotypes<a class="anchor" aria-label="anchor" href="#generating-genotypes-and-phenotypes"></a>
</h4>
<p>To generate both genotypes and phenotypes, we need to include a
<code>sim_mv</code> object that contains effect sizes. For our example,
we will use the data frame sample size format to indicate that three
individuals have both phenotypes measured and seven have only one or the
other measured. Note that we don’t need to include the <code>J</code>
argument because the <code>sim_mv</code> object contains all the
necessary information. The vector of <code>M</code> phenotypes for each
individual is simulated as <span class="math inline">\(Y_i = Y_{G,i} +
Y_{E,i}\)</span> where <span class="math inline">\(Y_{G,i}\)</span> is
the vector of genetic components computed from the genotypes and effect
sizes stored in the original <code>sim_mv</code> object and <span class="math inline">\(Y_{E,i}\)</span> is a vector of environmental
components sampled from a normal distribution with variance given by the
<code>Sigma_E</code> matrix in the original <code>sim_mv</code>
object.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="st">"trait_1"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>, <span class="st">"trait_2"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, <span class="st">"N"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">3</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">new_ind_dat1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/resample_inddata.html">resample_inddata</a></span><span class="op">(</span>N <span class="op">=</span> <span class="va">N</span>, </span>
<span>                                 dat <span class="op">=</span> <span class="va">orig_dat</span>, </span>
<span>                                 R_LD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">my_ld1</span><span class="op">)</span>, </span>
<span>                                 af <span class="op">=</span> <span class="va">af1</span>, </span>
<span>                                 calc_sumstats <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Generating both genotypes and phenotypes.</span></span>
<span><span class="co">#&gt; SNP effects provided for 12 SNPs and 2 traits.</span></span>
<span><span class="co">#&gt; I will assume that the environmental variance is the same in the old and new population.</span></span>
<span><span class="co">#&gt; I will assume that environmental correlation is the same in the old and new population. Note that this could result in different overall trait correlations.</span></span></code></pre></div>
<p>Genotype data are stored in <code>new_ind_dat1$X</code> and phenotype
data are in <code>new_ind_dat1$Y</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">new_ind_dat1</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "X"          "Y"          "af"         "Sigma_G"    "Sigma_E"   </span></span>
<span><span class="co">#&gt; [6] "pheno_sd"   "h2"         "trait_corr" "beta_joint"</span></span>
<span><span class="va">new_ind_dat1</span><span class="op">$</span><span class="va">X</span></span>
<span><span class="co">#&gt;     [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12]</span></span>
<span><span class="co">#&gt; S1     0    1    1    1    1    0    0    0    1     1     1     2</span></span>
<span><span class="co">#&gt; S2     1    1    1    2    2    1    1    1    1     1     1     1</span></span>
<span><span class="co">#&gt; S3     1    1    2    1    1    0    0    0    1     1     1     1</span></span>
<span><span class="co">#&gt; S4     0    0    0    1    1    1    1    1    1     1     1     1</span></span>
<span><span class="co">#&gt; S5     0    0    0    1    1    1    1    1    2     2     1     1</span></span>
<span><span class="co">#&gt; S6     0    0    0    2    2    1    1    0    2     2     2     1</span></span>
<span><span class="co">#&gt; S7     0    0    0    2    2    2    1    1    1     1     0     0</span></span>
<span><span class="co">#&gt; S8     0    0    0    1    1    0    0    1    2     2     2     2</span></span>
<span><span class="co">#&gt; S9     2    2    1    0    0    0    1    1    2     2     2     2</span></span>
<span><span class="co">#&gt; S10    0    1    1    2    2    1    1    1    1     1     2     1</span></span>
<span><span class="va">new_ind_dat1</span><span class="op">$</span><span class="va">Y</span></span>
<span><span class="co">#&gt;            y_1         y_2</span></span>
<span><span class="co">#&gt; 1  -0.64246004          NA</span></span>
<span><span class="co">#&gt; 2   1.26150568          NA</span></span>
<span><span class="co">#&gt; 3  -0.05425048          NA</span></span>
<span><span class="co">#&gt; 4   0.47374118  0.77563114</span></span>
<span><span class="co">#&gt; 5  -0.66257053 -0.64414580</span></span>
<span><span class="co">#&gt; 6   1.34406827 -0.08414463</span></span>
<span><span class="co">#&gt; 7           NA  0.42914109</span></span>
<span><span class="co">#&gt; 8           NA -2.21483702</span></span>
<span><span class="co">#&gt; 9           NA -1.61651866</span></span>
<span><span class="co">#&gt; 10          NA  0.55091258</span></span></code></pre></div>
<p>The returned object also includes some additional information
including <code>Sigma_G</code>, <code>Sigma_E</code>, and
<code>beta_joint</code> which have the same meaning as the objects with
the same name in a <code>sim_mv</code> object. If we had set
<code>calc_sumstats</code> to <code>TRUE</code>, the object would also
include effect size and standard error estimates for the association
between each variant and each trait.</p>
</div>
<div class="section level4">
<h4 id="generating-phenotypes-only">Generating Phenotypes Only<a class="anchor" aria-label="anchor" href="#generating-phenotypes-only"></a>
</h4>
<p>Finally, we can use <code>resample_inddata</code> to calculate
phenotypes for a previously generated set of genotypes from a
<code>sim_mv</code> object. To do this, we supply the genotypes matrix
to the <code>genos</code> argument. If this is supplied, no genotypes
matrix will be returned. Even though the genotype matrix is supplied, we
also need to supply the <code>R_LD</code> and <code>af</code> arguments
for the population LD and allele frequencies. These are used to
correctly calculate the genetic variance-covariance matrix. Note that it
is important that the total number of individuals implied by the sample
size argument match the number of rows in the genotypes matrix. The
following call will generate an error because we supplied 15 genotypes
but the sample size data frame only includes 10 individuals.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phenos_only</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/resample_inddata.html">resample_inddata</a></span><span class="op">(</span>N <span class="op">=</span> <span class="va">N</span>, </span>
<span>                                dat <span class="op">=</span> <span class="va">orig_dat</span>, </span>
<span>                                genos <span class="op">=</span> <span class="va">genos_only</span><span class="op">$</span><span class="va">X</span>,</span>
<span>                                R_LD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">my_ld1</span><span class="op">)</span>,</span>
<span>                                af <span class="op">=</span> <span class="va">af1</span>,</span>
<span>                                calc_sumstats <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Generating phenotypes only.</span></span>
<span><span class="co">#&gt; Error in check_matrix(genos, "genos", ntotal, J): Expected genos to have 10 rows, found 15</span></span></code></pre></div>
<p>We can fix the poblem by only including the first 10 rowsof the
genotype matrix.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phenos_only</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/resample_inddata.html">resample_inddata</a></span><span class="op">(</span>N <span class="op">=</span> <span class="va">N</span>, </span>
<span>                                dat <span class="op">=</span> <span class="va">orig_dat</span>, </span>
<span>                                genos <span class="op">=</span> <span class="va">genos_only</span><span class="op">$</span><span class="va">X</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="op">]</span>,</span>
<span>                                R_LD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">my_ld1</span><span class="op">)</span>,</span>
<span>                                af <span class="op">=</span> <span class="va">af1</span>,</span>
<span>                                calc_sumstats <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Generating phenotypes only.</span></span>
<span><span class="co">#&gt; SNP effects provided for 12 SNPs and 2 traits.</span></span>
<span><span class="co">#&gt; I will assume that the environmental variance is the same in the old and new population.</span></span>
<span><span class="co">#&gt; I will assume that environmental correlation is the same in the old and new population. Note that this could result in different overall trait correlations.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">phenos_only</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Y"          "af"         "Sigma_G"    "Sigma_E"    "pheno_sd"  </span></span>
<span><span class="co">#&gt; [6] "h2"         "trait_corr" "beta_joint"</span></span>
<span><span class="va">phenos_only</span><span class="op">$</span><span class="va">Y</span></span>
<span><span class="co">#&gt;            y_1        y_2</span></span>
<span><span class="co">#&gt; 1   1.22836882         NA</span></span>
<span><span class="co">#&gt; 2  -0.42435626         NA</span></span>
<span><span class="co">#&gt; 3  -0.62076763         NA</span></span>
<span><span class="co">#&gt; 4  -0.86060862  0.5125080</span></span>
<span><span class="co">#&gt; 5   0.09021464 -1.0949420</span></span>
<span><span class="co">#&gt; 6  -0.69423142 -1.4815503</span></span>
<span><span class="co">#&gt; 7           NA -1.1507476</span></span>
<span><span class="co">#&gt; 8           NA  0.1640274</span></span>
<span><span class="co">#&gt; 9           NA -0.4314887</span></span>
<span><span class="co">#&gt; 10          NA -0.8675082</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="resampling-data-from-a-different-population">Resampling Data from a Different Population<a class="anchor" aria-label="anchor" href="#resampling-data-from-a-different-population"></a>
</h2>
<p>Both <code>resample_sumstats</code> and <code>resample_inddata</code>
can be used as above with different LD and allele frequencies than used
to create the original object. However, there are some special
considerations about trait variance and scaling that we will discuss in
this section.</p>
<div class="section level3">
<h3 id="understanding-effect-size-units">Understanding Effect Size Units<a class="anchor" aria-label="anchor" href="#understanding-effect-size-units"></a>
</h3>
<p>To understand resampled data that use different LD or allele
frequencies than the original data, it is important to understand the
units of effect sizes produced by <code>sim_mv</code>.
<code>sim_mv</code> always assumes that phenotype variance is equal to
1, so the interpretation of the effects in <code>beta_joint</code> is
the change in phenotype in units of phenotype SD per change in genotype
in units of either allele or genotype SD. To see the genotype unit, we
can check the <code>geno_scale</code> object. For example,</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">orig_dat</span><span class="op">$</span><span class="va">geno_scale</span></span>
<span><span class="co">#&gt; [1] "allele"</span></span></code></pre></div>
<p>By default, <code>sim_mv</code> will use the per-allele scale if
allele frequencies are available and otherwise use the per-SD scale. We
can also check the the phenotypes are scaled to unit variance in
<code>pheno_sd</code> which will always be a vector of 1’s if the object
was produced by <code>sim_mv</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">orig_dat</span><span class="op">$</span><span class="va">pheno_sd</span></span>
<span><span class="co">#&gt; [1] 1 1</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="changing-ld-and-allele-frequencies">Changing LD and Allele Frequencies<a class="anchor" aria-label="anchor" href="#changing-ld-and-allele-frequencies"></a>
</h3>
<p>The resampling functions in <code>GWASBrewer</code> assume that
effect sizes in the new population are the same as in the old population
on the genotype and phenotype scale given. If you are changing
populations, it is much more sensible to assume that effect sizes are
constant on the per-allele scale than on the per-genotype SD scale. This
means that if you want to regenerate data from a different population,
it is a good idea to start with input data that have
<code>geno_scale</code> equal to “allele”.</p>
<p>One consequence of differing allele frequencies and LD structure is
that the total genetic variance will be different in the new population
than in the old. By default, the resampling functions will assume that
the environmental variance is the same in the two populations. This
means that the overall variance of the phenotype in the new population
will probably not be 1. The resampling functions will
<strong>not</strong> rescale the phenotype in the new population because
this would mean that the phenotypes in the two populations had different
units and were not comparable. However, you can rescale the output after
the fact using if you would like the phenotypes to have unit
variance.</p>
<p>We can see this in action by resampling summary statistics using
different allele frequencies and LD pattern than used originally.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">af2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">my_ld2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0.3</span>, nrow <span class="op">=</span> <span class="fl">5</span>, ncol <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">my_ld2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="va">new_dat2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/resample_sumstats.html">resample_sumstats</a></span><span class="op">(</span>dat <span class="op">=</span> <span class="va">orig_dat</span>, </span>
<span>                              N <span class="op">=</span> <span class="va">N</span>, </span>
<span>                              R_LD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">my_ld2</span><span class="op">)</span>, </span>
<span>                              af <span class="op">=</span> <span class="va">af2</span><span class="op">)</span></span>
<span><span class="co">#&gt; Genetic variance in the new population differs from the genetic variance in the old population.</span></span>
<span><span class="co">#&gt; I will assume that the environmental variance is the same in the old and new population.</span></span>
<span><span class="co">#&gt; I will assume that environmental correlation is the same in the old and new population. Note that this could result in different overall trait correlations.</span></span>
<span><span class="co">#&gt; Note that the phenotype in the new population has a different variance from the phenotype in the old population.</span></span>
<span><span class="co">#&gt; I will keep the phenotype on the same scale as the original data, so effect sizes in the old and new object are comparable. If you would like to rescale the phenotype to have variance 1, use rescale_sumstats.</span></span>
<span><span class="co">#&gt; SNP effects provided for 12 SNPs and 2 traits.</span></span></code></pre></div>
<p>Notice that the function produces a message to let us know that the
phenotype has a different variance in the new population. We can check
by looking at <code>pheno_sd</code>. The genetic covariance matrix,
heritability, and overall trait correlation will also be different.
However, the environmental covariance will be the same.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">new_dat2</span><span class="op">$</span><span class="va">pheno_sd</span></span>
<span><span class="co">#&gt; [1] 0.9649663 0.9164662</span></span>
<span><span class="va">orig_dat</span><span class="op">$</span><span class="va">pheno_sd</span></span>
<span><span class="co">#&gt; [1] 1 1</span></span>
<span><span class="va">new_dat2</span><span class="op">$</span><span class="va">h2</span></span>
<span><span class="co">#&gt; [1] 0.04683348 0.11545049</span></span>
<span><span class="va">orig_dat</span><span class="op">$</span><span class="va">h2</span></span>
<span><span class="co">#&gt; [1] 0.1124495 0.2570578</span></span>
<span><span class="va">new_dat2</span><span class="op">$</span><span class="va">Sigma_G</span></span>
<span><span class="co">#&gt;             [,1]        [,2]</span></span>
<span><span class="co">#&gt; [1,] 0.043609461 0.008093447</span></span>
<span><span class="co">#&gt; [2,] 0.008093447 0.096968051</span></span>
<span><span class="va">orig_dat</span><span class="op">$</span><span class="va">Sigma_G</span></span>
<span><span class="co">#&gt;            [,1]       [,2]</span></span>
<span><span class="co">#&gt; [1,] 0.11244947 0.01934268</span></span>
<span><span class="co">#&gt; [2,] 0.01934268 0.25705783</span></span>
<span><span class="va">new_dat2</span><span class="op">$</span><span class="va">Sigma_E</span></span>
<span><span class="co">#&gt;           [,1]      [,2]</span></span>
<span><span class="co">#&gt; [1,] 0.8875505 0.1736201</span></span>
<span><span class="co">#&gt; [2,] 0.1736201 0.7429422</span></span>
<span><span class="va">orig_dat</span><span class="op">$</span><span class="va">Sigma_E</span></span>
<span><span class="co">#&gt;           [,1]      [,2]</span></span>
<span><span class="co">#&gt; [1,] 0.8875505 0.1736201</span></span>
<span><span class="co">#&gt; [2,] 0.1736201 0.7429422</span></span></code></pre></div>
<p>Note that the features that are kept constant are the environmental
variance <code>Sigma_E</code> and the per-allele joint effects,
<code>beta_joint</code>.</p>
</div>
<div class="section level3">
<h3 id="changing-environmental-variance-or-covariance">Changing Environmental Variance or Covariance<a class="anchor" aria-label="anchor" href="#changing-environmental-variance-or-covariance"></a>
</h3>
<p>Although the default behavior is to keep <code>Sigma_E</code>
constant, we can change this using the arguments
<code>new_env_var</code>, <code>new_h2</code>, <code>new_R_E</code> and
<code>new_R_obs</code>. These are options for both
<code>resample_sumstats</code> and <code>resample_inddata</code>. The
<code>new_env_var</code> and <code>new_h2</code> arguments provide
different ways of specifying the environmental variance so only one of
these can be used at a time. <code>new_h2</code> gives the heritability
in the new population while <code>new_env_var</code> directly gives the
environmental variance. The parameters <code>new_R_E</code> and
<code>new_R_obs</code> are alternate ways of specifying environmental
correlation. <code>new_R_E</code> directly specifies the enviromental
correlation while <code>new_R_obs</code> gives the total trait
correlation. Only one of these two parameters can be supplied.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">new_R_E</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">new_R_E</span><span class="op">[</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">new_R_E</span><span class="op">[</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0.4</span></span>
<span><span class="va">new_dat3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/resample_sumstats.html">resample_sumstats</a></span><span class="op">(</span>dat <span class="op">=</span> <span class="va">orig_dat</span>, </span>
<span>                              N <span class="op">=</span> <span class="va">N</span>, </span>
<span>                              R_LD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">my_ld2</span><span class="op">)</span>, </span>
<span>                              af <span class="op">=</span> <span class="va">af2</span>,</span>
<span>                              new_env_var <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.9</span>, <span class="fl">1.3</span><span class="op">)</span>, </span>
<span>                              new_R_E <span class="op">=</span> <span class="va">new_R_E</span><span class="op">)</span></span>
<span><span class="co">#&gt; Genetic variance in the new population differs from the genetic variance in the old population.</span></span>
<span><span class="co">#&gt; Note that the phenotype in the new population has a different variance from the phenotype in the old population.</span></span>
<span><span class="co">#&gt; I will keep the phenotype on the same scale as the original data, so effect sizes in the old and new object are comparable. If you would like to rescale the phenotype to have variance 1, use rescale_sumstats.</span></span>
<span><span class="co">#&gt; SNP effects provided for 12 SNPs and 2 traits.</span></span>
<span><span class="va">new_dat3</span><span class="op">$</span><span class="va">Sigma_E</span></span>
<span><span class="co">#&gt;           [,1]      [,2]</span></span>
<span><span class="co">#&gt; [1,] 0.9000000 0.4326662</span></span>
<span><span class="co">#&gt; [2,] 0.4326662 1.3000000</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cov2cor</a></span><span class="op">(</span><span class="va">new_dat3</span><span class="op">$</span><span class="va">Sigma_E</span><span class="op">)</span></span>
<span><span class="co">#&gt;      [,1] [,2]</span></span>
<span><span class="co">#&gt; [1,]  1.0  0.4</span></span>
<span><span class="co">#&gt; [2,]  0.4  1.0</span></span>
<span><span class="va">new_dat3</span><span class="op">$</span><span class="va">h2</span></span>
<span><span class="co">#&gt; [1] 0.04621558 0.06941322</span></span></code></pre></div>
<p>Note that specifying <code>new_h2</code> will always result in
exactly the heritability specified. This is in contrast with the
<code>h2</code> argument of <code>sim_mv</code> which gives the expected
heritability.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">new_dat4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/resample_sumstats.html">resample_sumstats</a></span><span class="op">(</span>dat <span class="op">=</span> <span class="va">orig_dat</span>, </span>
<span>                              N <span class="op">=</span> <span class="va">N</span>, </span>
<span>                              R_LD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">my_ld2</span><span class="op">)</span>, </span>
<span>                              af <span class="op">=</span> <span class="va">af2</span>,</span>
<span>                              new_h2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.15</span>, <span class="fl">0.25</span><span class="op">)</span>, </span>
<span>                              new_R_E <span class="op">=</span> <span class="va">new_R_E</span><span class="op">)</span></span>
<span><span class="co">#&gt; Genetic variance in the new population differs from the genetic variance in the old population.</span></span>
<span><span class="co">#&gt; Note that the phenotype in the new population has a different variance from the phenotype in the old population.</span></span>
<span><span class="co">#&gt; I will keep the phenotype on the same scale as the original data, so effect sizes in the old and new object are comparable. If you would like to rescale the phenotype to have variance 1, use rescale_sumstats.</span></span>
<span><span class="co">#&gt; SNP effects provided for 12 SNPs and 2 traits.</span></span>
<span><span class="va">new_dat4</span><span class="op">$</span><span class="va">h2</span></span>
<span><span class="co">#&gt; [1] 0.15 0.25</span></span>
<span><span class="va">new_dat4</span><span class="op">$</span><span class="va">Sigma_E</span></span>
<span><span class="co">#&gt;           [,1]      [,2]</span></span>
<span><span class="co">#&gt; [1,] 0.2471203 0.1072480</span></span>
<span><span class="co">#&gt; [2,] 0.1072480 0.2909042</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="rescaling-effect-size-units">Rescaling Effect Size Units<a class="anchor" aria-label="anchor" href="#rescaling-effect-size-units"></a>
</h3>
<p>If for any reason we want to change the units of effects in an object
produced by <code>sim_mv</code> or <code>resmaple_sumstats</code>, we
can use <code>rescale_sumstats</code>. For example, in the code below,
we rescale the effects in <code>orig_dat</code> to be on the per-SD
scale. Note that by doing this, we will delete the allele frequency
information in the <code>snp_info</code> table. This is because allele
frequency is irrelevant for per-SD scaled effects. To convert back to
per-allele scale, we would need to supply the allele frequency.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">orig_rescale1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rescale_sumstats.html">rescale_sumstats</a></span><span class="op">(</span>dat <span class="op">=</span> <span class="va">orig_dat</span>, </span>
<span>                                  output_geno_scale <span class="op">=</span> <span class="st">"sd"</span><span class="op">)</span></span>
<span><span class="va">orig_rescale1</span><span class="op">$</span><span class="va">beta_joint</span></span>
<span><span class="co">#&gt;               [,1]          [,2]</span></span>
<span><span class="co">#&gt;  [1,]  0.000000000  0.0000000000</span></span>
<span><span class="co">#&gt;  [2,] -0.097136125  0.0188220320</span></span>
<span><span class="co">#&gt;  [3,]  0.000000000  0.0000000000</span></span>
<span><span class="co">#&gt;  [4,]  0.146918495  0.0293836989</span></span>
<span><span class="co">#&gt;  [5,]  0.000000000  0.0000000000</span></span>
<span><span class="co">#&gt;  [6,]  0.000000000  0.0061492991</span></span>
<span><span class="co">#&gt;  [7,] -0.003777444 -0.0007554889</span></span>
<span><span class="co">#&gt;  [8,]  0.000000000 -0.5047783058</span></span>
<span><span class="co">#&gt;  [9,]  0.000000000  0.0000000000</span></span>
<span><span class="co">#&gt; [10,]  0.000000000  0.0000000000</span></span>
<span><span class="co">#&gt; [11,]  0.000000000  0.0000000000</span></span>
<span><span class="co">#&gt; [12,] -0.280286489 -0.0560572978</span></span>
<span><span class="va">orig_rescale1</span><span class="op">$</span><span class="va">geno_scale</span></span>
<span><span class="co">#&gt; [1] "sd"</span></span>
<span><span class="va">orig_dat</span><span class="op">$</span><span class="va">beta_joint</span></span>
<span><span class="co">#&gt;               [,1]         [,2]</span></span>
<span><span class="co">#&gt;  [1,]  0.000000000  0.000000000</span></span>
<span><span class="co">#&gt;  [2,] -0.149884295  0.029043026</span></span>
<span><span class="co">#&gt;  [3,]  0.000000000  0.000000000</span></span>
<span><span class="co">#&gt;  [4,]  0.231374881  0.046274976</span></span>
<span><span class="co">#&gt;  [5,]  0.000000000  0.000000000</span></span>
<span><span class="co">#&gt;  [6,]  0.000000000  0.009116327</span></span>
<span><span class="co">#&gt;  [7,] -0.005828723 -0.001165745</span></span>
<span><span class="co">#&gt;  [8,]  0.000000000 -0.728584727</span></span>
<span><span class="co">#&gt;  [9,]  0.000000000  0.000000000</span></span>
<span><span class="co">#&gt; [10,]  0.000000000  0.000000000</span></span>
<span><span class="co">#&gt; [11,]  0.000000000  0.000000000</span></span>
<span><span class="co">#&gt; [12,] -0.432491442 -0.086498288</span></span>
<span></span>
<span><span class="co">## back to per-allele scale</span></span>
<span><span class="va">orig_rescale2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rescale_sumstats.html">rescale_sumstats</a></span><span class="op">(</span>dat <span class="op">=</span> <span class="va">orig_rescale1</span>, </span>
<span>                                  output_geno_scale <span class="op">=</span> <span class="st">"allele"</span>, </span>
<span>                                  af <span class="op">=</span> <span class="va">orig_dat</span><span class="op">$</span><span class="va">snp_info</span><span class="op">$</span><span class="va">AF</span><span class="op">)</span></span>
<span><span class="va">orig_rescale2</span><span class="op">$</span><span class="va">beta_joint</span></span>
<span><span class="co">#&gt;               [,1]         [,2]</span></span>
<span><span class="co">#&gt;  [1,]  0.000000000  0.000000000</span></span>
<span><span class="co">#&gt;  [2,] -0.149884295  0.029043026</span></span>
<span><span class="co">#&gt;  [3,]  0.000000000  0.000000000</span></span>
<span><span class="co">#&gt;  [4,]  0.231374881  0.046274976</span></span>
<span><span class="co">#&gt;  [5,]  0.000000000  0.000000000</span></span>
<span><span class="co">#&gt;  [6,]  0.000000000  0.009116327</span></span>
<span><span class="co">#&gt;  [7,] -0.005828723 -0.001165745</span></span>
<span><span class="co">#&gt;  [8,]  0.000000000 -0.728584727</span></span>
<span><span class="co">#&gt;  [9,]  0.000000000  0.000000000</span></span>
<span><span class="co">#&gt; [10,]  0.000000000  0.000000000</span></span>
<span><span class="co">#&gt; [11,]  0.000000000  0.000000000</span></span>
<span><span class="co">#&gt; [12,] -0.432491442 -0.086498288</span></span></code></pre></div>
<p>We can also change the scale of the outcomes.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">orig_rescale3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rescale_sumstats.html">rescale_sumstats</a></span><span class="op">(</span>dat <span class="op">=</span> <span class="va">orig_dat</span>, </span>
<span>                                  output_geno_scale <span class="op">=</span> <span class="st">"allele"</span>, </span>
<span>                                  output_pheno_sd <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.4</span>, <span class="fl">0.8</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>There are a few effects of changing the phenotype scale. First the
effect sizes are different. In this case the effect sizes for trait 1
have been multiplied by 1.4 and the effect sizes for trait 2 have been
multiplied by 0.8. Second, the genetic and environmental covariance
matrices have been scaled appropriately.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">orig_rescale3</span><span class="op">$</span><span class="va">Sigma_G</span></span>
<span><span class="co">#&gt;            [,1]       [,2]</span></span>
<span><span class="co">#&gt; [1,] 0.22040097 0.02166381</span></span>
<span><span class="co">#&gt; [2,] 0.02166381 0.16451701</span></span>
<span><span class="va">orig_rescale3</span><span class="op">$</span><span class="va">Sigma_E</span></span>
<span><span class="co">#&gt;           [,1]      [,2]</span></span>
<span><span class="co">#&gt; [1,] 1.7395990 0.1944545</span></span>
<span><span class="co">#&gt; [2,] 0.1944545 0.4754830</span></span>
<span><span class="va">orig_rescale3</span><span class="op">$</span><span class="va">Sigma_G</span> <span class="op">+</span> <span class="va">orig_rescale3</span><span class="op">$</span><span class="va">Sigma_E</span></span>
<span><span class="co">#&gt;           [,1]      [,2]</span></span>
<span><span class="co">#&gt; [1,] 1.9600000 0.2161183</span></span>
<span><span class="co">#&gt; [2,] 0.2161183 0.6400000</span></span></code></pre></div>
<p>Note that the heritability, <code>orig_rescale3$h2</code> has not
changed but is no longer equal to the diagonal of <code>Sigma_G</code>.
The trait correlation in <code>trait_corr</code> is also the same. The
final difference is in the total and direct trait effects matrix. In our
case, trait 1 has been multiplied by 1.4 and trait 2 was multiplied by
0.8. On the original scale, a one unit increase in trait 1 caused a 0.2
unit increase in trait 2. So on the new scale, a 1 unit increase in
trait 1 causes a <span class="math inline">\(0.2 \cdot (0.8/1.4) =
0.114\)</span> unit increase in trait 2.</p>
<p>Note that the causal relationship between the traits implies the
following relationship between effect sizes:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">orig_dat</span>, <span class="fu"><a href="https://rdrr.io/r/base/all.equal.html" class="external-link">all.equal</a></span><span class="op">(</span><span class="va">direct_SNP_effects_joint</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span> <span class="op">+</span> <span class="va">total_trait_effects</span><span class="op">[</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">]</span><span class="op">*</span><span class="va">direct_SNP_effects_joint</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, <span class="va">beta_joint</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p>This is the basis of methods like Mendelian randomization. We can see
that after scaling, this relationship is still true.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">orig_rescale3</span>, <span class="fu"><a href="https://rdrr.io/r/base/all.equal.html" class="external-link">all.equal</a></span><span class="op">(</span><span class="va">direct_SNP_effects_joint</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span> <span class="op">+</span> <span class="va">total_trait_effects</span><span class="op">[</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">]</span><span class="op">*</span><span class="va">direct_SNP_effects_joint</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, <span class="va">beta_joint</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jean Morrison.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
