<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="GWASBrewer">
<title>Simulating Data • GWASBrewer</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulating Data">
<meta property="og:description" content="GWASBrewer">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">GWASBrewer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.0.0237</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/effect_distribution.html">Controlling Effect Size Distributions</a>
    <a class="dropdown-item" href="../articles/resampling.html">Resampling and Re-Scaling Summary and Individual Level Data</a>
    <a class="dropdown-item" href="../articles/simulating_data.html">Simulating Data</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/jean997/GWASBrewer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<link href="simulating_data_files/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="simulating_data_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><script src="simulating_data_files/viz-1.8.2/viz.js"></script><link href="simulating_data_files/DiagrammeR-styles-0.2/styles.css" rel="stylesheet">
<script src="simulating_data_files/grViz-binding-1.0.11/grViz.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Simulating Data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jean997/GWASBrewer/blob/HEAD/vignettes/simulating_data.Rmd" class="external-link"><code>vignettes/simulating_data.Rmd</code></a></small>
      <div class="d-none name"><code>simulating_data.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jean997/GWASBrewer" class="external-link">GWASBrewer</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rich-iannone.github.io/DiagrammeR/" class="external-link">DiagrammeR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/hadley/reshape" class="external-link">reshape2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette demonstrates how to use the <code>sim_mv</code>
function to simulate data a few different types of GWAS data.</p>
</div>
<div class="section level2">
<h2 id="introduction-to-sim_mv">Introduction to <code>sim_mv</code><a class="anchor" aria-label="anchor" href="#introduction-to-sim_mv"></a>
</h2>
<p>The <code>sim_mv</code> function generates GWAS summary statistics
for multiple continuous traits from a linear structural equation model
encoded as a matrix of direct effects. Variants can be generated with or
without LD. There are also some helper functions for LD-pruning and
generating special kinds of direct effect matrices.</p>
<p>The <code>sim_mv</code> function internally calls a more general
function <code>sim_lf</code> which generates summary statistics given a
latent factor structure for a set of traits.</p>
<p>Throughout this vignette, we will use <span class="math inline">\(M\)</span> to represent the number of traits and
<span class="math inline">\(J\)</span> to represent the number of
variants.</p>
</div>
<div class="section level2">
<h2 id="basic-usage">Basic Usage<a class="anchor" aria-label="anchor" href="#basic-usage"></a>
</h2>
<div class="section level3">
<h3 id="input">Input<a class="anchor" aria-label="anchor" href="#input"></a>
</h3>
<p>The <code>sim_mv</code> function has five required arguments:</p>
<ul>
<li>
<code>N</code>: The GWAS sample size for each trait. This can be a
scalar, vector, matrix, or data frame. If a vector, <code>N</code>
should have length equal to <span class="math inline">\(M\)</span>. If
there are overlapping samples between GWAS, <code>N</code> should be an
<span class="math inline">\(M \times M\)</span> matrix or data frame
with format described in the “Sample Overlap” section below.</li>
<li>
<code>J</code>: The number of SNPs to simulate (scalar).</li>
<li>
<code>h2</code>: The heritability of each trait. This can be a
scalar or a length <span class="math inline">\(M\)</span> vector.</li>
<li>
<code>pi</code>: The proportion of SNPs that have a direct effect on
each trait. This can be a scalar, a length <span class="math inline">\(M\)</span> vector, or a matrix with <span class="math inline">\(J \times M\)</span> matrix (more details on the
matrix format can be found in the Effect Size Distribution
vignette).</li>
<li>
<code>G</code>: A matrix specifying direct causal effects between
traits. <code>G</code> should be an <span class="math inline">\(M\times
M\)</span> matrix described below. Alternatively, if there are no causal
effects between any traits, <code>G</code> can be given as the positive
integer <span class="math inline">\(M\)</span>. This is equivalent to
<code>G = matrix(0, nrow = M, ncol = M)</code>.</li>
</ul>
<p>There are additional optional arguments:</p>
<ul>
<li>
<code>R_E</code> and <code>R_obs</code>: Alternative ways to specify
environmental correlation (see “Sample Overlap” section below for more
details).</li>
<li>
<code>R_LD</code>: A list of LD blocks (see “Simulating Data with
LD”).</li>
<li>
<code>af</code>: Vector of allele frequencies (required if
<code>R_LD</code> is specified). If <code>R_LD</code> is not specified,
<code>af</code> can be a scalar, vector of length <code>J</code> or a
function (more details below). If <code>R_LD</code> is specified,
<code>af</code> must be a vector with length corresponding to the size
of the LD pattern (see “Simulating Data with LD”)</li>
<li>
<code>sporadic_pleiotropy</code>: Allow a single variant to have
direct effects on multiple traits at random. Defaults to
<code>TRUE</code>.</li>
<li>
<code>pi_exact</code>: If <code>TRUE</code>, the number of direct
effect SNPs for each trait will be exactly equal to
<code>round(pi*J)</code>. Defaults to <code>FALSE</code>.</li>
<li>
<code>h2_exact</code>: If <code>TRUE</code>, the heritability of
each trait will be exactly <code>h2</code>. Defaults to
<code>FALSE</code>.</li>
<li>
<code>est_s</code>: If <code>TRUE</code>, return estimates of
<code>se(beta_hat)</code>. Defaults to <code>FALSE</code> but we
generally recommend setting <code>est_s</code> to <code>TRUE</code> if
you will be making use of standard errors.</li>
<li>
<code>snp_effect_function</code> and <code>snp_info</code> are
parameters useful for specifying non-default distributions of effect
sizes. This is not covered in this vignette but is covered in the Effect
Size Distributions vignette.</li>
</ul>
</div>
<div class="section level3">
<h3 id="output">Output<a class="anchor" aria-label="anchor" href="#output"></a>
</h3>
<p>The <code>sim_mv</code> function returns an object with class
<code>sim_mv</code>. This object contains the following elements:</p>
<p>GWAS summary statistics are contained in two or three matrices:</p>
<ul>
<li>
<code>beta_hat</code>: Simulated GWAS effect estimates and standard
errors.</li>
<li>
<code>se_beta_hat</code>: True standard errors of
<code>beta_hat</code>.</li>
<li>
<code>s_estimate</code>: If <code>est_s= TRUE</code> then a
simulated estimate of <code>se_beta_hat</code>.</li>
</ul>
<p>True marginal and joint, total and direct effects are contained in
four matrices:</p>
<ul>
<li>
<code>beta_joint</code>: Total causal effects of SNPs on
traits.</li>
<li>
<code>beta_marg</code>: Expected marginal association of SNPs on
traits. <code>beta_marg</code> is the expected value of
<code>beta_hat</code>. When there is no LD, <code>beta_marg</code> and
<code>beta_joint</code> are the same.</li>
<li>
<code>direct_SNP_effects_joint</code>: Direct causal effects of SNPs
on traits. Direct means not mediated by other traits.</li>
<li>
<code>direct_SNP_effects_marg</code>: Like <code>beta_marg</code>
but considering only direct rather than total effects.</li>
</ul>
<p>The relationship between traits is contained in two matrices:</p>
<ul>
<li>
<code>direct_trait_effects</code>: Matrix of direct effects between
traits. This should be the same as the input <code>G</code>.</li>
<li>
<code>total_trait_effects</code>: Matrix of total effects between
traits.</li>
</ul>
<p>Trait covariance is described by four matrices:</p>
<ul>
<li>
<code>Sigma_G</code>: Genetic variance-covariance matrix, determined
by heritability and <code>G</code>.</li>
<li>
<code>Sigma_E</code>: Environmental variance-covariance matrix. This
is determined by heritability and <code>R_E</code>.</li>
<li>
<code>trait_corr</code>: Population trait correlation, equal to
<code>cov2cor(Sigma_G + Sigma_E)</code>. For data produces with
<code>sim_mv</code>, trait variance is always equal to 1, so
<code>trait_corr = Sigma_G + Sigma_E</code>.</li>
<li>
<code>R</code>: Correlation in sampling error of
<code>beta_hat</code> across traits, equal to <code>trait_corr</code>
scaled by a matrix of sample overlap proportions.</li>
</ul>
<p>Some other pieces of information useful in more complicated
scenarios:</p>
<ul>
<li>
<code>h2</code>: Realized trait heritability. For data produced with
<code>sim_mv</code>, this will always be equal to
<code>diag(Sigma_G)</code>. However this may not be the case for
resampled data (see the Resampling Data vignette).</li>
<li>
<code>pheno_sd</code>: Standard deviation of traits. For data
produced with <code>sim_mv</code>, this is always a vector of 1’s.</li>
<li>
<code>snp_info</code>: A data frame of variant information including
allele frequency and possibly other information (see the Effect
Distribution vignette).</li>
<li>
<code>geno_scale</code>: Equal to <code>allele</code> if effect
sizes are per allele or <code>sd</code> if effect sizes are per genotype
SD (i.e. standardized).</li>
</ul>
<p>The order of the columns of all results corresponds to the order of
traits in <code>G</code>.</p>
</div>
</div>
<div class="section level2">
<h2 id="simplest-usage">Simplest Usage<a class="anchor" aria-label="anchor" href="#simplest-usage"></a>
</h2>
<p>The simplest thing to do with <code>sim_mv</code> is to generate
summary statistics for <span class="math inline">\(M\)</span> traits
with no causal relationship with no LD between variants. In the code
below, we generate data for 3 traits and 100,000 variants. Since the
sample size is specified as a scalar, all three GWAS have the same
sample size but there are no overlapping samples. For some variety, we
give each trait a different heritability and a different proportion of
causal variants.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat_simple</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_mv.html">sim_mv</a></span><span class="op">(</span>G <span class="op">=</span> <span class="fl">3</span>, <span class="co"># using the shortcut for specifying unrelated traits.</span></span>
<span>                     <span class="co"># equivalent to G = matrix(0, nrow = 3, ncol = 3)</span></span>
<span>                     N <span class="op">=</span> <span class="fl">50000</span>, <span class="co"># sample size, same for all three traits</span></span>
<span>                     J <span class="op">=</span> <span class="fl">100000</span>, <span class="co"># number of variants</span></span>
<span>                     h2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.25</span>, <span class="fl">0.4</span><span class="op">)</span>, <span class="co"># heritability</span></span>
<span>                     pi <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.01</span>, <span class="fl">0.005</span>, <span class="fl">0.02</span><span class="op">)</span>, <span class="co"># proportion of causal variants</span></span>
<span>                     est_s <span class="op">=</span> <span class="cn">TRUE</span> <span class="co"># generate standard error estimates.</span></span>
<span>                     <span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 100000 SNPs and 3 traits.</span></span></code></pre></div>
<p>The <code>sim_mv</code> function always assumes that traits have been
scaled to have variance equal to 1, so effect sizes are interpretable as
the expected change in the trait in units of SD per either alternate
allele (if <code>geno_scale = allele</code>) or per genotype SD (if
<code>geno_scale = sd</code>). In our case,
<code>dat_simple$geno_scale</code> is equal to <code>sd</code> because
no allele frequencies were provided. The realized genetic and
environmental variance matrices are in <code>Sigma_G</code> and
<code>Sigma_E</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat_simple</span><span class="op">$</span><span class="va">Sigma_G</span></span>
<span><span class="co">#&gt;              [,1]         [,2]         [,3]</span></span>
<span><span class="co">#&gt; [1,] 1.045339e-01 4.515705e-06 0.0005262836</span></span>
<span><span class="co">#&gt; [2,] 4.515705e-06 2.530663e-01 0.0009271551</span></span>
<span><span class="co">#&gt; [3,] 5.262836e-04 9.271551e-04 0.4429666631</span></span>
<span><span class="va">dat_simple</span><span class="op">$</span><span class="va">Sigma_E</span></span>
<span><span class="co">#&gt;           [,1]      [,2]      [,3]</span></span>
<span><span class="co">#&gt; [1,] 0.8954661 0.0000000 0.0000000</span></span>
<span><span class="co">#&gt; [2,] 0.0000000 0.7469337 0.0000000</span></span>
<span><span class="co">#&gt; [3,] 0.0000000 0.0000000 0.5570333</span></span></code></pre></div>
<p>The diagonal of <code>dat_simple$Sigma_G</code> is equal to the trait
heritability (also stored in <code>dat_simple$h2</code>). We can see
that these numbers are close to but not exactly equal to the values
input to the <code>h2</code> parameter. This is because <code>h2</code>
provides the expected heritability. If we want to force the realized
heritability to be exactly equal to the input <code>h2</code>, we can
use <code>h2_exact = TRUE</code>. The genetic covariance (the
non-diagonal elements of <code>Sigma_G</code>) is slightly non-zero
because the traits share a small number of causal variants by chance. If
we want to prevent this, we can use
<code>sporadic_pleiotropy=FALSE</code>. However, in some cases, it is
not possible to satisfy <code>sporadic_pleiotropy = FALSE</code> and
that option will generate an error.</p>
<p>The simulated summary effect estimates are in
<code>dat_simple$beta_hat</code> and simulated standard error estimates
are in <code>dat_simple$s_estimate</code>. In our case, these will both
be 100,000 by 3 matrices.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dat_simple</span><span class="op">$</span><span class="va">beta_hat</span><span class="op">)</span></span>
<span><span class="co">#&gt;              [,1]         [,2]         [,3]</span></span>
<span><span class="co">#&gt; [1,] 0.0063610849 -0.007271609  0.003926946</span></span>
<span><span class="co">#&gt; [2,] 0.0009913487  0.012587585  0.003931295</span></span>
<span><span class="co">#&gt; [3,] 0.0031938057 -0.002387450 -0.003170087</span></span>
<span><span class="co">#&gt; [4,] 0.0030629600  0.001481305 -0.002369274</span></span>
<span><span class="co">#&gt; [5,] 0.0020414224  0.007953421  0.005325885</span></span>
<span><span class="co">#&gt; [6,] 0.0030602370 -0.002037463 -0.008673129</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dat_simple</span><span class="op">$</span><span class="va">s_estimate</span><span class="op">)</span></span>
<span><span class="co">#&gt;             [,1]        [,2]        [,3]</span></span>
<span><span class="co">#&gt; [1,] 0.004478273 0.004480611 0.004483377</span></span>
<span><span class="co">#&gt; [2,] 0.004468891 0.004478611 0.004482622</span></span>
<span><span class="co">#&gt; [3,] 0.004472107 0.004466985 0.004465830</span></span>
<span><span class="co">#&gt; [4,] 0.004465952 0.004460529 0.004476516</span></span>
<span><span class="co">#&gt; [5,] 0.004449150 0.004493462 0.004482853</span></span>
<span><span class="co">#&gt; [6,] 0.004480817 0.004464715 0.004479402</span></span></code></pre></div>
<p>In actual GWAS data, these two matrices are the only information we
get to observe. Everything else stored in <code>dat_simple</code> is
information that is unobservable in “real” data but useful for
benchmarking analysis methods. The effect estimates in
<code>dat_simple$beta_hat</code> are always estimates of the true
marginal effects in in <code>dat_simple$beta_marg</code>. Since there is
no LD in this data, marginal and joint effects are the same, so you will
find that <code>dat_simple$beta_marg</code> and
<code>dat_simple$beta_joint</code> are identical. We can identify causal
variants as variants with non-zero values of <code>beta_joint</code>.
For example, <code>which(dat_simple$beta_joint[,1] != 0)</code> would
give the indices of variants causal for trait 1.<br>
The estimates in <code>dat_simple$s_estimate</code> are estimates of the
standard error of <code>dat_simple$beta_hat</code>. The true standard
errors are stored in <code>dat_simple$se_beta_hat</code>. If we had left
<code>est_s</code> at its default value of <code>FALSE</code>,
<code>dat_simple</code> would not contain the <code>s_estimate</code>
matrix.</p>
</div>
<div class="section level2">
<h2 id="specifying-causal-relationships-between-traits">Specifying Causal Relationships Between Traits<a class="anchor" aria-label="anchor" href="#specifying-causal-relationships-between-traits"></a>
</h2>
<p>The matrix <code>G</code> specifies a linear structural equation
model for a set of traits. To generate a set of <span class="math inline">\(M\)</span> traits with no causal relationships,
<code>G</code> can be set either equal to <code>M</code> or to an <span class="math inline">\(M\times M\)</span> matrix of 0’s. Otherwise,
<code>G</code> must be an <span class="math inline">\(M \times
M\)</span> matrix with <code>G[i,j]</code> specifying the direct linear
effect of trait <span class="math inline">\(i\)</span> on trait <span class="math inline">\(j\)</span>. The diagonal entries of <span class="math inline">\(G\)</span> should be 0 (no self effects). An error
will be generated if <code>G</code> specifies a graph that contains
cycles. Since all traits have variance equal to 1, so
<code>G[i,j]^2</code> is the proportion of trait <span class="math inline">\(j\)</span> variance explained by the direct effect
of trait <span class="math inline">\(i\)</span>.</p>
<p>For example, the matrix</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">G</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">0.25</span><span class="op">)</span>, <span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">0.15</span><span class="op">)</span>, </span>
<span>              <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">0.1</span><span class="op">)</span>, </span>
<span>              <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">0.2</span><span class="op">)</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fl">0.3</span><span class="op">)</span>, </span>
<span>              <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">4</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">G</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="va">G</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"X"</span>, <span class="st">"Y"</span>, <span class="st">"Z"</span>, <span class="st">"W"</span><span class="op">)</span></span>
<span><span class="va">G</span></span>
<span><span class="co">#&gt;           X   Y Z          W</span></span>
<span><span class="co">#&gt; X 0.0000000 0.5 0  0.3872983</span></span>
<span><span class="co">#&gt; Y 0.0000000 0.0 0  0.3162278</span></span>
<span><span class="co">#&gt; Z 0.4472136 0.0 0 -0.5477226</span></span>
<span><span class="co">#&gt; W 0.0000000 0.0 0  0.0000000</span></span></code></pre></div>
<p>corresponds to the graph</p>
<div class="grViz html-widget html-fill-item" id="htmlwidget-ac96cb3ee4656e2e9ec3" style="width:480px;height:432.632880098888px;"></div>
<script type="application/json" data-for="htmlwidget-ac96cb3ee4656e2e9ec3">{"x":{"diagram":"digraph {\n\ngraph [layout = \"neato\",\n       outputorder = \"edgesfirst\",\n       bgcolor = \"white\"]\n\nnode [fontname = \"Helvetica\",\n      fontsize = \"10\",\n      shape = \"circle\",\n      fixedsize = \"true\",\n      width = \"0.5\",\n      style = \"filled\",\n      fillcolor = \"aliceblue\",\n      color = \"gray70\",\n      fontcolor = \"gray50\"]\n\nedge [fontname = \"Helvetica\",\n     fontsize = \"8\",\n     len = \"1.5\",\n     color = \"gray80\",\n     arrowsize = \"0.5\"]\n\n  \"1\" [label = \"X\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fillcolor = \"#FFFFFF\", fontcolor = \"black\", color = \"black\", pos = \"0,0!\"] \n  \"2\" [label = \"Y\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fillcolor = \"#FFFFFF\", fontcolor = \"black\", color = \"black\", pos = \"1,-0.5!\"] \n  \"3\" [label = \"Z\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fillcolor = \"#FFFFFF\", fontcolor = \"black\", color = \"black\", pos = \"1,1!\"] \n  \"4\" [label = \"W\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fillcolor = \"#FFFFFF\", fontcolor = \"black\", color = \"black\", pos = \"2,0!\"] \n\"3\"->\"1\" [color = \"black\", label = \"0.447\"] \n\"1\"->\"2\" [color = \"black\", label = \"0.5\"] \n\"1\"->\"4\" [color = \"black\", label = \"0.387\"] \n\"2\"->\"4\" [color = \"black\", label = \"0.316\"] \n\"3\"->\"4\" [color = \"black\", label = \"-0.548\"] \n}","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script><p>To simulate simple data from this graph, we can use</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim_dat1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_mv.html">sim_mv</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>,</span>
<span>                  N <span class="op">=</span> <span class="fl">50000</span>,</span>
<span>                  J <span class="op">=</span> <span class="fl">100000</span>,</span>
<span>                  h2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">0.3</span>, <span class="fl">0.5</span>, <span class="fl">0.4</span><span class="op">)</span>, </span>
<span>                  pi <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>                  est_s <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 100000 SNPs and 4 traits.</span></span></code></pre></div>
<p>In this specification, we have four GWAS with sample size 50,000,
again with no overlapping samples. Since <code>J = 100000</code> and
<code>pi = 0.01</code>, we expect each trait to have 1000 direct effect
variants.</p>
<p>With causal relationships between traits, there are now some
interesting things to notice about the output. First, we can see that
there is both genetic and environmental covariance between the
traits.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim_dat1</span><span class="op">$</span><span class="va">Sigma_G</span></span>
<span><span class="co">#&gt;            X          Y          Z           W</span></span>
<span><span class="co">#&gt; X 0.28249614 0.14135064  0.2179663  0.03485498</span></span>
<span><span class="co">#&gt; Y 0.14135064 0.29912608  0.1089297  0.08886324</span></span>
<span><span class="co">#&gt; Z 0.21796635 0.10892974  0.4892269 -0.14942217</span></span>
<span><span class="co">#&gt; W 0.03485498 0.08886324 -0.1494222  0.40289524</span></span>
<span><span class="va">sim_dat1</span><span class="op">$</span><span class="va">Sigma_E</span></span>
<span><span class="co">#&gt;           X         Y          Z          W</span></span>
<span><span class="co">#&gt; X 0.7175039 0.3545701  0.2288111  0.2619020</span></span>
<span><span class="co">#&gt; Y 0.3545701 0.7008739  0.1130720  0.2951464</span></span>
<span><span class="co">#&gt; Z 0.2288111 0.1130720  0.5107731 -0.1531603</span></span>
<span><span class="co">#&gt; W 0.2619020 0.2951464 -0.1531603  0.5971048</span></span>
<span><span class="va">sim_dat1</span><span class="op">$</span><span class="va">trait_corr</span></span>
<span><span class="co">#&gt;           X         Y          Z          W</span></span>
<span><span class="co">#&gt; X 1.0000000 0.4959207  0.4467774  0.2967570</span></span>
<span><span class="co">#&gt; Y 0.4959207 1.0000000  0.2220017  0.3840097</span></span>
<span><span class="co">#&gt; Z 0.4467774 0.2220017  1.0000000 -0.3025824</span></span>
<span><span class="co">#&gt; W 0.2967570 0.3840097 -0.3025824  1.0000000</span></span></code></pre></div>
<p>By default, <code>sim_mv</code> assumes that direct environmental
components of each trait are independent, meaning that the DAG explains
all of the correlation between traits. This is modifiable using the
<code>R_E</code> and <code>R_obs</code> arguments, discussed a bit later
in this vignette.</p>
<p>Relationships between traits are described by two matrices,
<code>direct_trait_effects</code> which is equal to the input
<code>G</code> matrix and <code>total_trait_effects</code> which gives
the total effect of each trait on each other trait. Here we can notice
that the direct effect of <span class="math inline">\(Z\)</span> on
<span class="math inline">\(W\)</span> is -0.548 as specified but the
total effect includes the effects mediated by <span class="math inline">\(X\)</span> and <span class="math inline">\(Y\)</span> as well, <span class="math inline">\(-0.548 + 0.447\cdot 0.387 + 0.447\cdot 0.5 \cdot
0.316 = -0.304\)</span>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim_dat1</span><span class="op">$</span><span class="va">direct_trait_effects</span></span>
<span><span class="co">#&gt;           X   Y Z          W</span></span>
<span><span class="co">#&gt; X 0.0000000 0.5 0  0.3872983</span></span>
<span><span class="co">#&gt; Y 0.0000000 0.0 0  0.3162278</span></span>
<span><span class="co">#&gt; Z 0.4472136 0.0 0 -0.5477226</span></span>
<span><span class="co">#&gt; W 0.0000000 0.0 0  0.0000000</span></span>
<span><span class="va">sim_dat1</span><span class="op">$</span><span class="va">total_trait_effects</span></span>
<span><span class="co">#&gt;           X         Y Z          W</span></span>
<span><span class="co">#&gt; X 0.0000000 0.5000000 0  0.5454122</span></span>
<span><span class="co">#&gt; Y 0.0000000 0.0000000 0  0.3162278</span></span>
<span><span class="co">#&gt; Z 0.4472136 0.2236068 0 -0.3038068</span></span>
<span><span class="co">#&gt; W 0.0000000 0.0000000 0  0.0000000</span></span></code></pre></div>
<p>We can also use the output to understand which variants have direct
effects on each trait and which have indirect (mediated) effects. The
<code>direct_SNP_effects_joint</code> object gives the direct effect of
each variant on each trait while <code>beta_joint</code> gives the the
total effect of each variant. Direct and total marginal effects are
stored in <code>direct_SNP_effects_marg</code> and
<code>beta_marg</code>. Since we again have no LD, the
<code>_marg</code> and <code>_joint</code> matrices are the same.</p>
<p>Direct SNP effects are always independent across traits while total
SNP effects are the sum of direct effects and indirect effects mediated
by other traits. We will make some plots to see the difference.</p>
<p>First we plot direct SNP effects on <span class="math inline">\(Z\)</span> vs direct SNP effects on <span class="math inline">\(W\)</span></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sim_dat1</span><span class="op">$</span><span class="va">direct_SNP_effects_joint</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span>, <span class="va">sim_dat1</span><span class="op">$</span><span class="va">direct_SNP_effects_joint</span><span class="op">[</span>,<span class="fl">4</span><span class="op">]</span>, </span>
<span>     xlab <span class="op">=</span> <span class="st">"Direct Z effect"</span>, ylab <span class="op">=</span> <span class="st">"Direct W effect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="simulating_data_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p>Most variants have direct effects on at most one of <span class="math inline">\(Z\)</span> or <span class="math inline">\(W\)</span> but a small number affect both because
<code>sporadic_pleiotropy = TRUE</code> by default.</p>
<p>Next we plot the total SNP effects on <span class="math inline">\(Z\)</span> vs the total SNP effects on <span class="math inline">\(W\)</span>. Because <span class="math inline">\(Z\)</span> has a causal effect on <span class="math inline">\(W\)</span>, all variants with effects on <span class="math inline">\(Z\)</span> also affect <span class="math inline">\(W\)</span>. The line in the plot has slope equal
to the total effect of <span class="math inline">\(Z\)</span> on <span class="math inline">\(W\)</span>. The majority of SNPs that have
non-zero effect on <span class="math inline">\(Z\)</span> fall exactly
on this line. With <code>sporadic_pleiotropy= FALSE</code>, all of the
variants with non-zero effect on <span class="math inline">\(Z\)</span>
would fall on this line. The variants on the vertical line at 0 are
variants with non-zero direct effect on <span class="math inline">\(W\)</span> but no direct effect on <span class="math inline">\(Z\)</span>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sim_dat1</span><span class="op">$</span><span class="va">beta_joint</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span>, <span class="va">sim_dat1</span><span class="op">$</span><span class="va">beta_joint</span><span class="op">[</span>,<span class="fl">4</span><span class="op">]</span>, </span>
<span>     xlab <span class="op">=</span> <span class="st">"Total Z effect"</span>, ylab <span class="op">=</span> <span class="st">"Total W effect"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">sim_dat1</span><span class="op">$</span><span class="va">total_trait_effects</span><span class="op">[</span><span class="fl">3</span>,<span class="fl">4</span><span class="op">]</span>, col <span class="op">=</span> <span class="st">"red"</span>, lty <span class="op">=</span> <span class="fl">2</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="simulating_data_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="specifying-allele-frequencies">Specifying Allele Frequencies<a class="anchor" aria-label="anchor" href="#specifying-allele-frequencies"></a>
</h2>
<p>Allele frequencies can be specified using the <code>af</code>
argument which can accept a scalar, a vector of length <span class="math inline">\(J\)</span>, or a function that takes a single
argument and returns a vector of allele frequencies with length
determined by the argument. If <code>af</code> is a scalar, the same
allele frequency is used for all variants. The function specification is
used in the example below. If the <code>af</code> argument is provided,
<code>sim_mv</code> will return all results on the per-allele scale and
the <code>geno_scale</code> element of the returned object will be equal
to <code>allele</code>. The <code>snp_info</code> element of the
returned object will also include the allele frequency of each
variant.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim_dat2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_mv.html">sim_mv</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>,</span>
<span>                  N <span class="op">=</span> <span class="fl">50000</span>, </span>
<span>                  J <span class="op">=</span> <span class="fl">10000</span>, </span>
<span>                  h2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">0.3</span>, <span class="fl">0.5</span>, <span class="fl">0.4</span><span class="op">)</span>, </span>
<span>                  pi <span class="op">=</span> <span class="fl">0.01</span>, </span>
<span>                  af <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/Beta.html" class="external-link">rbeta</a></span><span class="op">(</span><span class="va">n</span>, <span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 10000 SNPs and 4 traits.</span></span>
<span></span>
<span><span class="va">sim_dat2</span><span class="op">$</span><span class="va">geno_scale</span></span>
<span><span class="co">#&gt; [1] "allele"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sim_dat2</span><span class="op">$</span><span class="va">snp_info</span><span class="op">)</span></span>
<span><span class="co">#&gt;   SNP         AF</span></span>
<span><span class="co">#&gt; 1   1 0.05448172</span></span>
<span><span class="co">#&gt; 2   2 0.52558863</span></span>
<span><span class="co">#&gt; 3   3 0.02522715</span></span>
<span><span class="co">#&gt; 4   4 0.29653930</span></span>
<span><span class="co">#&gt; 5   5 0.29400173</span></span>
<span><span class="co">#&gt; 6   6 0.09964018</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="simulating-data-with-ld">Simulating Data with LD<a class="anchor" aria-label="anchor" href="#simulating-data-with-ld"></a>
</h2>
<p>The <code>sim_mv</code> function can be used to generate data with LD
by inputting a list of LD matrices and corresponding allele frequency
information. The function will work fastest if the LD matrix is broken
into smallish independent blocks. The input data format for the LD
pattern is a list of either a) matrices, b) sparse matrices (class
<code>dsCMatrix</code>) or c) eigen decompositions (class
<code>eigen</code>). <code>R_LD</code> is interpreted as providing
blocks in a block-diagonal SNP correlation matrix.</p>
<p>Importantly, the supplied LD pattern does not have to be the same
size as the number of SNPs we wish to generate (<code>J</code>). It will
be repeated or subsetted as necessary to create an LD pattern of the
appropriate size.</p>
<p>The package contains a built-in data set containing the LD pattern
from Chromosome 19 in HapMap3 broken into 39 blocks. This LD pattern was
estimated from the HapMap3 European subset using LDShrink. This data set
can also be downloaded <a href="https://zenodo.org/record/6761943#.Yrno2njMIUE" class="external-link">here</a>. The LD
pattern must be accompanied by a vector of allele frequencies with
length equal to the total size of the LD pattern (i.e. the sum of the
size of each block in the list).</p>
<p>Let’s look at the built-in LD data</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"ld_mat_list"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"AF"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">ld_mat_list</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 39</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">ld_mat_list</span>, <span class="va">class</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix"</span></span>
<span><span class="co">#&gt;  [7] "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix"</span></span>
<span><span class="co">#&gt; [13] "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix"</span></span>
<span><span class="co">#&gt; [19] "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix"</span></span>
<span><span class="co">#&gt; [25] "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix"</span></span>
<span><span class="co">#&gt; [31] "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix" "dsCMatrix"</span></span>
<span><span class="co">#&gt; [37] "dsCMatrix" "dsCMatrix" "dsCMatrix"</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This prints the number of SNPs in each block</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">ld_mat_list</span>, <span class="va">nrow</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] 140 519 339 435 523 280 675 325 651 548 274 483 442 744 460 177 469 173 358</span></span>
<span><span class="co">#&gt; [20] 564 392 737 596 818 307 863 276 435 204 364 480 381 757 844 753 656 483 856</span></span>
<span><span class="co">#&gt; [39] 709</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">ld_mat_list</span>, <span class="va">nrow</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 19490</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">AF</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 19490</span></span></code></pre></div>
<p>The LD pattern covers 19,490 SNPs, equal to the length of the
<code>AF</code> vector. The built-in LD pattern corresponds to a density
of about 1.2 million variants per genome. However, for this example, we
will generate data for only 100k variants. This means that causal
effects will be denser than they might be in more realistic data with
the same number of effect variants.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">sim_dat1_LD</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_mv.html">sim_mv</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>,</span>
<span>                      J <span class="op">=</span> <span class="fl">1e5</span>, </span>
<span>                      N <span class="op">=</span> <span class="fl">50000</span>, </span>
<span>                      h2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">0.3</span>, <span class="fl">0.5</span>, <span class="fl">0.4</span><span class="op">)</span>, </span>
<span>                      pi <span class="op">=</span> <span class="fl">0.01</span>, </span>
<span>                      R_LD <span class="op">=</span> <span class="va">ld_mat_list</span>, </span>
<span>                      af <span class="op">=</span> <span class="va">AF</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 100000 SNPs and 4 traits.</span></span></code></pre></div>
<p>In data with LD, the <code>_joint</code> objects and
<code>_marg</code> objects are not identical. For example, we can
compare <code>beta_joint</code> and <code>beta_marg</code> for the third
trait (<span class="math inline">\(Z\)</span>).</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">sim_dat1_LD</span>, <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">beta_joint</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span>, <span class="va">beta_marg</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, lty <span class="op">=</span> <span class="fl">2</span>, lwd <span class="op">=</span><span class="fl">2</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<p><img src="simulating_data_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p>Variants with non-zero values of <code>beta_joint[,3]</code> have
causal effects on <span class="math inline">\(Z\)</span> while those
with non-zero values of <code>beta_marg[,3]</code> have non-zero
marginal association with <span class="math inline">\(Z\)</span>,
meaning that they are in LD with at least one causal variant. In the
plot, we see that many variants with no causal effect have non-zero
marginal association, which is expected. The causal variants don’t fall
exactly on the red line because, multiple causal variants may fall into
the same LD block.</p>
<div class="section level3">
<h3 id="ld-pruning-ld-proxies-and-ld-matrix-extraction">LD-Pruning, LD-Proxies, and LD Matrix Extraction<a class="anchor" aria-label="anchor" href="#ld-pruning-ld-proxies-and-ld-matrix-extraction"></a>
</h3>
<p>Many post-GWAS applications such as Mendelian randomization and
polygenic risk score construction require an LD-pruned set of variants.
<code>GWASBrewer</code> contains a few LD-related functions to help with
pruning and testing methods that require input LD matrices. Note that
all of these methods use the true LD pattern rather than estimated
LD.</p>
<p>The <code>sim_ld_prune</code> function will perform LD-clumping on
simulated data, prioritizing variants according to a supplied
<code>pvalue</code> vector. Although this argument is called
<code>pvalue</code>, it can be any numeric vector used to prioritize
variants. The <code>pvalue</code> argument can also accept an integer.
If <code>pvalue = i</code>, variants will be prioritized according to
the p-value for the <code>i</code>th trait in the simulated data. If
<code>pvalue</code> is omitted, variants will be prioritized randomly
(so a different result will be obtained each re-run unless a seed is
set).</p>
<p>To speed up performance, if you only need variants with <span class="math inline">\(p\)</span>-value less than a certain threshold,
supply the <code>pvalue_thresh</code> argument. Below we prune based on
the p-values for trait <span class="math inline">\(Z\)</span> in two
equivalent ways.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pruned_set1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_ld_prune.html">sim_ld_prune</a></span><span class="op">(</span>dat <span class="op">=</span> <span class="va">sim_dat1_LD</span>, </span>
<span>                            pvalue <span class="op">=</span> <span class="fl">3</span>, </span>
<span>                            R_LD <span class="op">=</span> <span class="va">ld_mat_list</span>, </span>
<span>                            r2_thresh <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                            pval_thresh <span class="op">=</span> <span class="fl">1e-6</span><span class="op">)</span></span>
<span><span class="co">#&gt; Prioritizing variants based on p-value for trait 3</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">pruned_set1</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 354</span></span>
<span><span class="va">pval3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">sim_dat1_LD</span>, <span class="fl">2</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">pnorm</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">beta_hat</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span><span class="op">/</span><span class="va">se_beta_hat</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pruned_set2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_ld_prune.html">sim_ld_prune</a></span><span class="op">(</span>dat <span class="op">=</span> <span class="va">sim_dat1_LD</span>, </span>
<span>                            pvalue <span class="op">=</span> <span class="va">pval3</span>, </span>
<span>                            R_LD <span class="op">=</span> <span class="va">ld_mat_list</span>, </span>
<span>                            r2_thresh <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                            pval_thresh <span class="op">=</span> <span class="fl">1e-6</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html" class="external-link">all.equal</a></span><span class="op">(</span><span class="va">pruned_set1</span>, <span class="va">pruned_set2</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p><code>sim_ld_prune</code> returns a vector of indices corresponding
to an LD-pruned set of variants.</p>
<p>The <code>sim_ld_proxy</code> function will return indices of
LD-proxies (variants with LD above a given threshold) with a supplied
set of variants. Here we extract proxies for a few arbitrary variants.
The <code>return_mat</code> option will cause the function to return the
LD matrix for the proxies as well as the indices of proxies</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ld_proxies</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_ld_proxy.html">sim_ld_proxy</a></span><span class="op">(</span><span class="va">sim_dat1_LD</span>, index <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">400</span>, <span class="fl">600</span><span class="op">)</span>, R_LD <span class="op">=</span> <span class="va">ld_mat_list</span>, r2_thresh <span class="op">=</span> <span class="fl">0.64</span>, return_mat <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ld_proxies</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; [[1]]$index</span></span>
<span><span class="co">#&gt; [1] 100</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[1]]$proxy_index</span></span>
<span><span class="co">#&gt; [1] 98 99</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[1]]$Rproxy</span></span>
<span><span class="co">#&gt;           100        98        99</span></span>
<span><span class="co">#&gt; 100 1.0000000 0.9568668 0.9745924</span></span>
<span><span class="co">#&gt; 98  0.9568668 1.0000000 0.9798682</span></span>
<span><span class="co">#&gt; 99  0.9745924 0.9798682 1.0000000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; [[2]]$index</span></span>
<span><span class="co">#&gt; [1] 400</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]$proxy_index</span></span>
<span><span class="co">#&gt; [1] 395 396 397 398 399 401 402</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]$Rproxy</span></span>
<span><span class="co">#&gt;           400       395       396       397       398       399       401</span></span>
<span><span class="co">#&gt; 400 1.0000000 0.9495092 0.9661446 0.9793188 0.9730871 0.9306611 0.9105497</span></span>
<span><span class="co">#&gt; 395 0.9495092 1.0000000 0.9812294 0.9675572 0.9742288 0.9302869 0.8876271</span></span>
<span><span class="co">#&gt; 396 0.9661446 0.9812294 1.0000000 0.9841205 0.9905250 0.9454487 0.9054550</span></span>
<span><span class="co">#&gt; 397 0.9793188 0.9675572 0.9841205 1.0000000 0.9810106 0.9367627 0.9175545</span></span>
<span><span class="co">#&gt; 398 0.9730871 0.9742288 0.9905250 0.9810106 1.0000000 0.9522425 0.9119614</span></span>
<span><span class="co">#&gt; 399 0.9306611 0.9302869 0.9454487 0.9367627 0.9522425 1.0000000 0.8731959</span></span>
<span><span class="co">#&gt; 401 0.9105497 0.8876271 0.9054550 0.9175545 0.9119614 0.8731959 1.0000000</span></span>
<span><span class="co">#&gt; 402 0.8080771 0.7821419 0.7934351 0.8097815 0.7991366 0.7643812 0.7522628</span></span>
<span><span class="co">#&gt;           402</span></span>
<span><span class="co">#&gt; 400 0.8080771</span></span>
<span><span class="co">#&gt; 395 0.7821419</span></span>
<span><span class="co">#&gt; 396 0.7934351</span></span>
<span><span class="co">#&gt; 397 0.8097815</span></span>
<span><span class="co">#&gt; 398 0.7991366</span></span>
<span><span class="co">#&gt; 399 0.7643812</span></span>
<span><span class="co">#&gt; 401 0.7522628</span></span>
<span><span class="co">#&gt; 402 1.0000000</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt; [[3]]$index</span></span>
<span><span class="co">#&gt; [1] 600</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]$proxy_index</span></span>
<span><span class="co">#&gt; [1] 601 606</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]$Rproxy</span></span>
<span><span class="co">#&gt;            600        601        606</span></span>
<span><span class="co">#&gt; 600  1.0000000 -0.8044757  0.8670863</span></span>
<span><span class="co">#&gt; 601 -0.8044757  1.0000000 -0.7193148</span></span>
<span><span class="co">#&gt; 606  0.8670863 -0.7193148  1.0000000</span></span></code></pre></div>
<p>Finally, the <code>sim_extract_ld</code> function will extract the LD
matrix for a set of variants.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ld_mat1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_extract_ld.html">sim_extract_ld</a></span><span class="op">(</span><span class="va">sim_dat1_LD</span>, index <span class="op">=</span> <span class="fl">600</span><span class="op">:</span><span class="fl">606</span>, R_LD <span class="op">=</span> <span class="va">ld_mat_list</span><span class="op">)</span></span>
<span><span class="va">ld_mat1</span></span>
<span><span class="co">#&gt;            600        601        602        603        604        605</span></span>
<span><span class="co">#&gt; 600  1.0000000 -0.8044757 -0.6715969 -0.7920287 -0.7323562 -0.4605856</span></span>
<span><span class="co">#&gt; 601 -0.8044757  1.0000000  0.8354364  0.9791149  0.9404418  0.6067155</span></span>
<span><span class="co">#&gt; 602 -0.6715969  0.8354364  1.0000000  0.8138201  0.7891286  0.6555054</span></span>
<span><span class="co">#&gt; 603 -0.7920287  0.9791149  0.8138201  1.0000000  0.9363088  0.5849580</span></span>
<span><span class="co">#&gt; 604 -0.7323562  0.9404418  0.7891286  0.9363088  1.0000000  0.5890082</span></span>
<span><span class="co">#&gt; 605 -0.4605856  0.6067155  0.6555054  0.5849580  0.5890082  1.0000000</span></span>
<span><span class="co">#&gt; 606  0.8670863 -0.7193148 -0.5968298 -0.7200977 -0.7469780 -0.3902207</span></span>
<span><span class="co">#&gt;            606</span></span>
<span><span class="co">#&gt; 600  0.8670863</span></span>
<span><span class="co">#&gt; 601 -0.7193148</span></span>
<span><span class="co">#&gt; 602 -0.5968298</span></span>
<span><span class="co">#&gt; 603 -0.7200977</span></span>
<span><span class="co">#&gt; 604 -0.7469780</span></span>
<span><span class="co">#&gt; 605 -0.3902207</span></span>
<span><span class="co">#&gt; 606  1.0000000</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="specifying-sample-size-sample-overlap-and-environmental-correlation">Specifying Sample Size, Sample Overlap, and Environmental
Correlation<a class="anchor" aria-label="anchor" href="#specifying-sample-size-sample-overlap-and-environmental-correlation"></a>
</h2>
<p>If two GWAS are performed on different traits using overlapping
samples, the sampling errors of effect estimates will be correlated. If
the two GWAS have sample sizes <span class="math inline">\(N_1\)</span>
and <span class="math inline">\(N_2\)</span> with <span class="math inline">\(N_c\)</span> overlapping samples, then the
correlation of <span class="math inline">\(\hat{z}_{1j}\)</span> and
<span class="math inline">\(\hat{z}_{2j}\)</span>, <span class="math inline">\(z\)</span>-scores for variant <span class="math inline">\(j\)</span> in study 1 and study 2, is
approximately <span class="math inline">\(\frac{N_c}{\sqrt{N_1 N_2}}
\rho_{1,2}\)</span> where <span class="math inline">\(\rho_{1,2}\)</span> is the observational trait
correlation (assuming the studies are conducted in the same super
population). Below we describe how to specify the observational
correlation and sample overlap.</p>
<div class="section level3">
<h3 id="specifying-sample-size-and-sample-overlap">Specifying Sample Size and Sample Overlap<a class="anchor" aria-label="anchor" href="#specifying-sample-size-and-sample-overlap"></a>
</h3>
<p>The sample size argument, <code>N</code>, can be specified as a
scalar, a vector, a matrix, or a data frame. Both the scalar and matrix
specification indicate no overlapping samples between GWAS. To specify
sample overlap, we need to use the matrix or data frame formats. If
<code>N</code> is a matrix, it should have dimension <span class="math inline">\(M\times M\)</span> with <code>N[i,i]</code> giving
the sample size of study <span class="math inline">\(i\)</span> and
<code>N[i,j]</code> giving the number of samples that are in both study
<span class="math inline">\(i\)</span> and study <span class="math inline">\(j\)</span>. In data frame format, <span class="math inline">\(N\)</span> should have columns named
<code>trait_1</code>, … <code>trait_[M]</code> and <code>N</code>. The
<code>trait_[x]</code> columns will be interpreted as logicals and the
<code>N</code> column should give the number of samples in each
combination of studies. For example, the following specifications for
two traits are equivalent.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">60000</span>, <span class="fl">30000</span>, <span class="fl">30000</span>, <span class="fl">60000</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">N</span></span>
<span><span class="co">#&gt;       [,1]  [,2]</span></span>
<span><span class="co">#&gt; [1,] 60000 30000</span></span>
<span><span class="co">#&gt; [2,] 30000 60000</span></span>
<span></span>
<span><span class="va">Ndf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>trait_1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span>, </span>
<span>                  trait_2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, </span>
<span>                  N <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">30000</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">Ndf</span></span>
<span><span class="co">#&gt;   trait_1 trait_2     N</span></span>
<span><span class="co">#&gt; 1       1       0 30000</span></span>
<span><span class="co">#&gt; 2       1       1 30000</span></span>
<span><span class="co">#&gt; 3       0       1 30000</span></span></code></pre></div>
<p>When there are more than two traits, the data frame format contains
more information than the matrix format. This format is required by the
<code>resample_inddata</code> function (covered in a different
vignette). For <code>sim_mv</code>, either format is sufficient.</p>
</div>
<div class="section level3">
<h3 id="using-sample-size-0-to-omit-traits">Using Sample Size 0 to Omit Traits<a class="anchor" aria-label="anchor" href="#using-sample-size-0-to-omit-traits"></a>
</h3>
<p>In some circumstances, we may want to generate true effects for a
trait but not generate summary statistics. In this case, we can use a
sample size of 0. Setting <code>N = 0</code> will mean that
<code>beta_hat</code>, <code>s_estimate</code>, <code>se_beta_hat</code>
will be <code>NA</code> for all traits. Alternatively, if <code>N</code>
is a vector with some 0 elements, only those traits will be missing.
Finally, if <code>N</code> is a matrix, we can set the row and column
corresponding to the omitted trait to zero. For example, in the
specification below, we omit summary statistics for <span class="math inline">\(Z\)</span>, the third trait. We also specify some
overlapping samples in studies for the other three traits.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">50000</span>, <span class="fl">10000</span>, <span class="fl">0</span>, <span class="fl">0</span>, </span>
<span>              <span class="fl">10000</span>, <span class="fl">40000</span>, <span class="fl">0</span>, <span class="fl">10000</span>, </span>
<span>              <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, <span class="fl">0</span>, </span>
<span>              <span class="fl">0</span>, <span class="fl">10000</span>, <span class="fl">0</span>, <span class="fl">20000</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">N</span></span>
<span><span class="co">#&gt;       [,1]  [,2] [,3]  [,4]</span></span>
<span><span class="co">#&gt; [1,] 50000 10000    0     0</span></span>
<span><span class="co">#&gt; [2,] 10000 40000    0 10000</span></span>
<span><span class="co">#&gt; [3,]     0     0    0     0</span></span>
<span><span class="co">#&gt; [4,]     0 10000    0 20000</span></span>
<span><span class="va">sim_dat2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_mv.html">sim_mv</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>,</span>
<span>                  N <span class="op">=</span> <span class="va">N</span>,</span>
<span>                  J <span class="op">=</span> <span class="fl">100000</span>,</span>
<span>                  h2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">0.3</span>, <span class="fl">0.5</span>, <span class="fl">0.4</span><span class="op">)</span>, </span>
<span>                  pi <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>                  est_s <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 100000 SNPs and 4 traits.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sim_dat2</span><span class="op">$</span><span class="va">beta_hat</span><span class="op">)</span></span>
<span><span class="co">#&gt;               [,1]         [,2] [,3]          [,4]</span></span>
<span><span class="co">#&gt; [1,]  1.258828e-03 -0.002383374   NA  1.074629e-02</span></span>
<span><span class="co">#&gt; [2,] -8.580051e-03 -0.015482893   NA -7.281355e-03</span></span>
<span><span class="co">#&gt; [3,] -2.515356e-03 -0.001070368   NA -2.189328e-02</span></span>
<span><span class="co">#&gt; [4,]  5.422213e-05 -0.003164621   NA  8.103383e-03</span></span>
<span><span class="co">#&gt; [5,] -8.150978e-03 -0.005807492   NA  5.922625e-05</span></span>
<span><span class="co">#&gt; [6,] -5.861632e-03 -0.003012872   NA  3.302201e-03</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sim_dat2</span><span class="op">$</span><span class="va">se_beta_hat</span><span class="op">)</span></span>
<span><span class="co">#&gt;             [,1]  [,2] [,3]        [,4]</span></span>
<span><span class="co">#&gt; [1,] 0.004472136 0.005   NA 0.007071068</span></span>
<span><span class="co">#&gt; [2,] 0.004472136 0.005   NA 0.007071068</span></span>
<span><span class="co">#&gt; [3,] 0.004472136 0.005   NA 0.007071068</span></span>
<span><span class="co">#&gt; [4,] 0.004472136 0.005   NA 0.007071068</span></span>
<span><span class="co">#&gt; [5,] 0.004472136 0.005   NA 0.007071068</span></span>
<span><span class="co">#&gt; [6,] 0.004472136 0.005   NA 0.007071068</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sim_dat2</span><span class="op">$</span><span class="va">s_estimate</span><span class="op">)</span></span>
<span><span class="co">#&gt;             [,1]        [,2] [,3]        [,4]</span></span>
<span><span class="co">#&gt; [1,] 0.004453566 0.004976698   NA 0.007050114</span></span>
<span><span class="co">#&gt; [2,] 0.004471810 0.005008104   NA 0.007081267</span></span>
<span><span class="co">#&gt; [3,] 0.004471716 0.004976063   NA 0.007033821</span></span>
<span><span class="co">#&gt; [4,] 0.004459041 0.004982154   NA 0.007068942</span></span>
<span><span class="co">#&gt; [5,] 0.004480086 0.004961544   NA 0.007006829</span></span>
<span><span class="co">#&gt; [6,] 0.004472674 0.004987358   NA 0.007045390</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="understanding-genetic-and-environmental-covariance">Understanding Genetic and Environmental Covariance<a class="anchor" aria-label="anchor" href="#understanding-genetic-and-environmental-covariance"></a>
</h3>
<p>In the model used by <code>GWASBrewer</code>, each trait has a direct
genetic component, a direct environmental component, and components from
effects of other traits in the DAG. For example, in the four trait DAG
we have been working with, the underlying model really looks like
this:</p>
<div class="grViz html-widget html-fill-item" id="htmlwidget-e5c8c404fe174e4c81bd" style="width:480px;height:432.632880098888px;"></div>
<script type="application/json" data-for="htmlwidget-e5c8c404fe174e4c81bd">{"x":{"diagram":"digraph {\n\ngraph [layout = \"neato\",\n       outputorder = \"edgesfirst\",\n       bgcolor = \"white\"]\n\nnode [fontname = \"Helvetica\",\n      fontsize = \"10\",\n      shape = \"circle\",\n      fixedsize = \"true\",\n      width = \"0.5\",\n      style = \"filled\",\n      fillcolor = \"aliceblue\",\n      color = \"gray70\",\n      fontcolor = \"gray50\"]\n\nedge [fontname = \"Helvetica\",\n     fontsize = \"8\",\n     len = \"1.5\",\n     color = \"gray80\",\n     arrowsize = \"0.5\"]\n\n  \"1\" [label = \"X\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fillcolor = \"#FFFFFF\", fontcolor = \"black\", color = \"black\", pos = \"0,0!\"] \n  \"2\" [label = \"Y\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fillcolor = \"#FFFFFF\", fontcolor = \"black\", color = \"black\", pos = \"1,-0.5!\"] \n  \"3\" [label = \"Z\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fillcolor = \"#FFFFFF\", fontcolor = \"black\", color = \"black\", pos = \"1,1!\"] \n  \"4\" [label = \"W\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fillcolor = \"#FFFFFF\", fontcolor = \"black\", color = \"black\", pos = \"2,0!\"] \n  \"5\" [label = \"Gx\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fillcolor = \"#FFFFFF\", fontcolor = \"black\", color = \"blue\", pos = \"-0.5,0.25!\"] \n  \"6\" [label = \"Ex\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fillcolor = \"#FFFFFF\", fontcolor = \"black\", color = \"blue\", pos = \"-0.5,-0.25!\"] \n  \"7\" [label = \"Gz\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fillcolor = \"#FFFFFF\", fontcolor = \"black\", color = \"blue\", pos = \"0.8,1.5!\"] \n  \"8\" [label = \"Ez\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fillcolor = \"#FFFFFF\", fontcolor = \"black\", color = \"blue\", pos = \"1.2,1.5!\"] \n  \"9\" [label = \"Gy\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fillcolor = \"#FFFFFF\", fontcolor = \"black\", color = \"blue\", pos = \"0.8,-1!\"] \n  \"10\" [label = \"Ey\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fillcolor = \"#FFFFFF\", fontcolor = \"black\", color = \"blue\", pos = \"1.2,-1!\"] \n  \"11\" [label = \"Gw\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fillcolor = \"#FFFFFF\", fontcolor = \"black\", color = \"blue\", pos = \"2.5,0.25!\"] \n  \"12\" [label = \"Ew\", fontname = \"Helvetica\", fontsize = \"10\", width = \"0.3\", fillcolor = \"#FFFFFF\", fontcolor = \"black\", color = \"blue\", pos = \"2.5,-0.25!\"] \n\"3\"->\"1\" [color = \"black\", label = \"0.447\"] \n\"1\"->\"2\" [color = \"black\", label = \"0.5\"] \n\"1\"->\"4\" [color = \"black\", label = \"0.387\"] \n\"2\"->\"4\" [color = \"black\", label = \"0.316\"] \n\"3\"->\"4\" [color = \"black\", label = \"-0.548\"] \n\"5\"->\"1\" [color = \"black\", color = \"black\"] \n\"6\"->\"1\" [color = \"black\", color = \"black\"] \n\"7\"->\"3\" [color = \"black\", color = \"black\"] \n\"8\"->\"3\" [color = \"black\", color = \"black\"] \n\"9\"->\"2\" [color = \"black\", color = \"black\"] \n\"10\"->\"2\" [color = \"black\", color = \"black\"] \n\"11\"->\"4\" [color = \"black\", color = \"black\"] \n\"12\"->\"4\" [color = \"black\", color = \"black\"] \n}","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script><p>In this picture, <span class="math inline">\(G_x\)</span>, <span class="math inline">\(G_y\)</span>, <span class="math inline">\(G_z\)</span>, and <span class="math inline">\(G_w\)</span> are direct genetic components of each
trait and <span class="math inline">\(E_x\)</span>, <span class="math inline">\(E_y\)</span>, <span class="math inline">\(E_z\)</span>, and <span class="math inline">\(E_w\)</span> are direct environmental components.
We always assume that the direct genetic components are independent of
each other and of the environmental components. By default, we also
assume that the environmental components are independent of each other
(so all blue circles in the picture above are mutually independent).
This means that, by default, the observational trait correlation is
explained completely by the specified DAG. In this case, the default
trait correlation is</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim_dat1</span><span class="op">$</span><span class="va">trait_corr</span></span>
<span><span class="co">#&gt;           X         Y          Z          W</span></span>
<span><span class="co">#&gt; X 1.0000000 0.4959207  0.4467774  0.2967570</span></span>
<span><span class="co">#&gt; Y 0.4959207 1.0000000  0.2220017  0.3840097</span></span>
<span><span class="co">#&gt; Z 0.4467774 0.2220017  1.0000000 -0.3025824</span></span>
<span><span class="co">#&gt; W 0.2967570 0.3840097 -0.3025824  1.0000000</span></span></code></pre></div>
<p>The simulation data object contains four matrices that describe trait
covarince, <code>Sigma_E</code>, the total environmental trait
covariance, <code>Sigma_G</code> the total genetic trait covariance,
<code>trait_corr</code>, the trait correlation equal to
<code>Sigma_G + Sigma_E</code>, and <code>R</code>, the row correlation
of <code>beta_hat</code>. <code>R</code> is equal to
<code>trait_corr</code> scaled by the a matrix of sample overlap
proportions.</p>
<p><code>Sigma_G</code> is always determined by the DAG and the
heritabilities. Currently, there are two ways to modify
<code>Sigma_E</code>. The first is to specify <code>R_obs</code> which
directly specifies the observational trait correlation
(<code>trait_corr</code>). In some cases, it is possible to request an
observational correlation matrix that is impossible. For example, in our
example, <span class="math inline">\(Z\)</span> has a strong negative
effect on <span class="math inline">\(W\)</span> so it is not possible
that all four traits are mutually strongly positively correlated. We can
get an error using</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">R_obs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0.8</span>, nrow <span class="op">=</span> <span class="fl">4</span>, ncol <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">R_obs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span></span>
<span><span class="va">wont_run</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_mv.html">sim_mv</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>,</span>
<span>                   J <span class="op">=</span> <span class="fl">50000</span>,</span>
<span>                  N <span class="op">=</span> <span class="fl">60000</span>,  </span>
<span>                  h2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">0.3</span>, <span class="fl">0.5</span>, <span class="fl">0.4</span><span class="op">)</span>, </span>
<span>                  pi <span class="op">=</span> <span class="fl">1000</span><span class="op">/</span><span class="fl">50000</span>, </span>
<span>                  R_obs <span class="op">=</span> <span class="va">R_obs</span> <span class="op">)</span></span>
<span><span class="co">#&gt; Error in value[[3L]](cond): R_obs is incompatible with trait relationships and heritability.</span></span></code></pre></div>
<p>A quick way to find out if your desired observational correlation is
feasible is to compute <code>R_obs - Sigma_G</code> and check that this
matrix is positive definite (i.e. check that it has all positive
eigenvalues).</p>
<p>An alternative way to specify the environmental correlation is to
specify <code>R_E</code> which gives the total correlation of
environmental components, i.e. <code>cov2cor(Sigma_E)</code>.
Importantly, <code>R_E</code> is the correlation of the
<strong>total</strong> environmental components, not the direct
environmental components shown in the graph above. Any positive definite
correlation matrix is a valid input for <code>R_E</code>, so we could
use</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">R_E</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0.8</span>, nrow <span class="op">=</span> <span class="fl">4</span>, ncol <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">R_E</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span></span>
<span><span class="va">sim_dat3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sim_mv.html">sim_mv</a></span><span class="op">(</span>G <span class="op">=</span> <span class="va">G</span>,</span>
<span>                   J <span class="op">=</span> <span class="fl">50000</span>,</span>
<span>                  N <span class="op">=</span> <span class="fl">60000</span>,  </span>
<span>                  h2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">0.3</span>, <span class="fl">0.5</span>, <span class="fl">0.4</span><span class="op">)</span>, </span>
<span>                  pi <span class="op">=</span> <span class="fl">1000</span><span class="op">/</span><span class="fl">50000</span>, </span>
<span>                  R_E <span class="op">=</span> <span class="va">R_E</span><span class="op">)</span></span>
<span><span class="co">#&gt; SNP effects provided for 50000 SNPs and 4 traits.</span></span></code></pre></div>
<p>which results in a total trait correlation of</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim_dat3</span><span class="op">$</span><span class="va">trait_corr</span></span>
<span><span class="co">#&gt;           X         Y         Z         W</span></span>
<span><span class="co">#&gt; X 1.0000000 0.7109713 0.7020261 0.5677883</span></span>
<span><span class="co">#&gt; Y 0.7109713 1.0000000 0.5966303 0.6110510</span></span>
<span><span class="co">#&gt; Z 0.7020261 0.5966303 1.0000000 0.3158510</span></span>
<span><span class="co">#&gt; W 0.5677883 0.6110510 0.3158510 1.0000000</span></span></code></pre></div>
<p>It is not currently possible to specify the correlation of the direct
environmental components.</p>
<p>Note that the environmental covariance and observational trait
correlation only influence the distribution of summary statistics if
there is overlap between GWAS samples. This means that
<code>sim_dat3</code> in the previous code block is actually a sample
from exactly the same distribution as <code>sim_dat1</code>, because our
specification has no sample overlap. We can tell that this is the case
because <code>sim_dat4$R</code> is the identity, indicating that all
summary statistics are independent across traits.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim_dat3</span><span class="op">$</span><span class="va">R</span></span>
<span><span class="co">#&gt;      [,1] [,2] [,3] [,4]</span></span>
<span><span class="co">#&gt; [1,]    1    0    0    0</span></span>
<span><span class="co">#&gt; [2,]    0    1    0    0</span></span>
<span><span class="co">#&gt; [3,]    0    0    1    0</span></span>
<span><span class="co">#&gt; [4,]    0    0    0    1</span></span></code></pre></div>
<p>By contrast, in <code>sim_dat2</code>, we did have sample overlap, so
there is non-zero correlation between the summary statistics.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim_dat2</span><span class="op">$</span><span class="va">R</span></span>
<span><span class="co">#&gt;           [,1]      [,2] [,3]      [,4]</span></span>
<span><span class="co">#&gt; [1,] 1.0000000 0.1120147    0 0.0000000</span></span>
<span><span class="co">#&gt; [2,] 0.1120147 1.0000000    0 0.1355998</span></span>
<span><span class="co">#&gt; [3,] 0.0000000 0.0000000    0 0.0000000</span></span>
<span><span class="co">#&gt; [4,] 0.0000000 0.1355998    0 1.0000000</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jean Morrison.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
