<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Sample individual level data with joint effects matching a sim_mv object"><title>Sample individual level data with joint effects matching a sim_mv object — resample_inddata • simGWAS</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Sample individual level data with joint effects matching a sim_mv object — resample_inddata"><meta property="og:description" content="Sample individual level data with joint effects matching a sim_mv object"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">simGWAS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0.0124</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/effect_distribution.html">Controlling Effect Size Distributions</a>
    <a class="dropdown-item" href="../articles/resampling.html">Resampling and Re-Scaling Summary and Individual Level Data</a>
    <a class="dropdown-item" href="../articles/simulating_data.html">Simulating Data</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/jean997/simGWAS/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Sample individual level data with joint effects matching a sim_mv object</h1>
      <small class="dont-index">Source: <a href="https://github.com/jean997/simGWAS/blob/HEAD/R/resample_inddata.R" class="external-link"><code>R/resample_inddata.R</code></a></small>
      <div class="d-none name"><code>resample_inddata.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Sample individual level data with joint effects matching a sim_mv object</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">resample_inddata</span><span class="op">(</span></span>
<span>  <span class="va">N</span>,</span>
<span>  dat <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  genos <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  J <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  R_LD <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  af <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sim_func <span class="op">=</span> <span class="va">gen_genos_mvn</span>,</span>
<span>  new_env_var <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  new_h2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  new_R_E <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  new_R_obs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  calc_sumstats <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>N</dt>
<dd><p>Sample size, scalar, vector, or special sample size format data frame, see details.</p></dd>


<dt>dat</dt>
<dd><p>An object of class <code>sim_mv</code> (produced by <code>sim_mv</code>). If `dat` is omitted,
the function will generate a matrix of genotypes only. If `dat` is provided, phenotypes for the traits
in `dat` will also be included.</p></dd>


<dt>genos</dt>
<dd><p>Optional matrix of pre-generated genotypes. If <code>genos</code> is supplied, <code>resample_inddata</code> will only generate phenotypes.</p></dd>


<dt>J</dt>
<dd><p>Optional number of variants. <code>J</code> is only required if <code>dat</code> is missing.</p></dd>


<dt>R_LD</dt>
<dd><p>LD pattern (optional). See <code><a href="sim_mv.html">?sim_mv</a></code> for more details.</p></dd>


<dt>af</dt>
<dd><p>Allele frequencies. <code>af</code> is required unless unless <code>genos</code> is supplied.</p></dd>


<dt>new_env_var</dt>
<dd><p>Optional. The environmental variance in the new population.
If missing the function will assume the environmental variance is the same as in the old population.</p></dd>


<dt>new_h2</dt>
<dd><p>Optional. The heritability in the new population. Provide at most one of <code>new_env_var</code> and <code>new_h2</code>.</p></dd>


<dt>new_R_E</dt>
<dd><p>Optional, specify environmental correlation in the new population.
If missing, the function will assume the environmental correlation is the same as in the original data.</p></dd>


<dt>new_R_obs</dt>
<dd><p>Optional, specify overall trait correlation in the new population. Specify at most one of <code>new_R_E</code> or <code>new_R_obs</code>.
If missing, the function will assume the environmental correlation is the same as in the original data.</p></dd>


<dt>calc_sumstats</dt>
<dd><p>If <code>TRUE</code>, associations between genotypes and phenotypes will be calculated and returned.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function can be used to generate individual level genotype and phenotype data. It can be used in three modes:</p>
<p>To generate genotype data only: No <code>sim_mv</code> object needs to be included. Supply only <code>N</code> as a single integer for the
number of individuals, <code>J</code> for the number of variants, <code>af</code>, and <code>R_LD</code> if desired. All other
parameters are not relevant if there is no phenotype, so if they are supplied, you will get an error. The returned object will include a
<code>N x J</code> matrix of genotypes and a vector of allele frequencies.</p>
<p>To generate both genotype and phenotype data: Supply <code>dat</code> (a <code>sim_mv</code> object) and leave <code>genos</code> missing. <code>N</code> and <code>af</code>
are required and all other options are optional.</p>
<p>To generate phenotype data only: Supply <code>dat</code> (a <code>sim_mv</code> object) and provide a matrix of genotypes to the <code>genos</code> argument. The number of
rows in <code>genos</code> must be equal to the total number of individuals implied by <code>N</code>.
So for example, if there are two traits with 10 samples each and no overlap, <code>genos</code>
should have 20 rows. The <code>R_LD</code> and <code>af</code> arguments should contain the population
LD and allele frequencies used to produce the genotypes. These are used to compute the genetic variance-covariance matrix.
<code>N</code> and <code>af</code> are required and all other options are optional.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Use resample_inddata to generate genotypes only</span></span></span>
<span class="r-in"><span><span class="va">simple_ld</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0.5</span>, nrow <span class="op">=</span> <span class="fl">5</span>, ncol <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">simple_ld</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span></span></span>
<span class="r-in"><span><span class="va">genos_only</span> <span class="op">&lt;-</span> <span class="fu">resample_inddata</span><span class="op">(</span>N <span class="op">=</span> <span class="fl">8</span>,</span></span>
<span class="r-in"><span>                               J <span class="op">=</span> <span class="fl">20</span>,</span></span>
<span class="r-in"><span>                               R_LD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">simple_ld</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                               af <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: hapsim</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: MASS</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Generating genotype matrix only.</span>
<span class="r-in"><span><span class="co"># generate genotypes and phenotypes</span></span></span>
<span class="r-in"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu"><a href="sim_mv.html">sim_mv</a></span><span class="op">(</span>N <span class="op">=</span> <span class="fl">0</span>,</span></span>
<span class="r-in"><span>              G <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>              J <span class="op">=</span> <span class="fl">20</span>,</span></span>
<span class="r-in"><span>              pi <span class="op">=</span> <span class="fl">0.5</span>,</span></span>
<span class="r-in"><span>              h2 <span class="op">=</span> <span class="fl">0.05</span>,</span></span>
<span class="r-in"><span>              R_LD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">simple_ld</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>              af <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> SNP effects provided for 20 SNPs and 1 traits.</span>
<span class="r-in"><span><span class="va">genos_and_phenos</span> <span class="op">&lt;-</span> <span class="fu">resample_inddata</span><span class="op">(</span>dat <span class="op">=</span> <span class="va">dat</span>,</span></span>
<span class="r-in"><span>                                      N <span class="op">=</span> <span class="fl">8</span>,</span></span>
<span class="r-in"><span>                                      R_LD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">simple_ld</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                                      af <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Generating both genotypes and phenotypes.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> SNP effects provided for 20 SNPs and 1 traits.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> I will assume that the environmental variance is the same in the old and new population.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> I will assume that environmental correlation is the same in the old and new population. Note that this could result in different overall trait correlations.</span>
<span class="r-in"><span><span class="co"># generate phenos only</span></span></span>
<span class="r-in"><span><span class="va">phenos_only</span> <span class="op">&lt;-</span> <span class="fu">resample_inddata</span><span class="op">(</span>dat <span class="op">=</span> <span class="va">dat</span>,</span></span>
<span class="r-in"><span>                                genos <span class="op">=</span> <span class="va">genos_only</span><span class="op">$</span><span class="va">X</span>,</span></span>
<span class="r-in"><span>                                N <span class="op">=</span> <span class="fl">8</span>,</span></span>
<span class="r-in"><span>                                R_LD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">simple_ld</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                                af <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0.3</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Generating phenotypes only.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> SNP effects provided for 20 SNPs and 1 traits.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> I will assume that the environmental variance is the same in the old and new population.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> I will assume that environmental correlation is the same in the old and new population. Note that this could result in different overall trait correlations.</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Jean Morrison.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

