<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Simulate multivariate GWAS data"><title>Simulate multivariate GWAS data — sim_mv • GWASBrewer</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulate multivariate GWAS data — sim_mv"><meta property="og:description" content="Simulate multivariate GWAS data"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">GWASBrewer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.0.0237</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/effect_distribution.html">Controlling Effect Size Distributions</a>
    <a class="dropdown-item" href="../articles/resampling.html">Resampling and Re-Scaling Summary and Individual Level Data</a>
    <a class="dropdown-item" href="../articles/simulating_data.html">Simulating Data</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/jean997/GWASBrewer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Simulate multivariate GWAS data</h1>
      <small class="dont-index">Source: <a href="https://github.com/jean997/GWASBrewer/blob/HEAD/R/sim_mv.R" class="external-link"><code>R/sim_mv.R</code></a></small>
      <div class="d-none name"><code>sim_mv.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Simulate multivariate GWAS data</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">sim_mv</span><span class="op">(</span></span>
<span>  <span class="va">N</span>,</span>
<span>  <span class="va">J</span>,</span>
<span>  <span class="va">h2</span>,</span>
<span>  <span class="va">pi</span>,</span>
<span>  G <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  est_s <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  R_obs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  R_E <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  R_LD <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  R_LD_eigen <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  af <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  snp_effect_function <span class="op">=</span> <span class="st">"normal"</span>,</span>
<span>  snp_info <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sporadic_pleiotropy <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  pi_exact <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  h2_exact <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>N</dt>
<dd><p>GWAS sample size. N can be a scalar, vector, or matrix. If N is a scalar, all GWAS have the same sample size
and there is no overlap between studies. If N is a vector, each element of N specifies the sample size of the corresponding
GWAS and there is no overlap between studies. If N is a matrix, N_ii specifies the sample size of study i
and N_ij specifies the number of samples present in both study i and study j. The elements of N must be positive
but non-integer values will not generate an error.</p></dd>


<dt>J</dt>
<dd><p>Number of variants to simulate</p></dd>


<dt>h2</dt>
<dd><p>A scalar or vector giving the heritability of each trait.</p></dd>


<dt>pi</dt>
<dd><p>A scalar or vector giving the expected proportion of direct effect SNPs for each trait.</p></dd>


<dt>G</dt>
<dd><p>Matrix of direct effects. Rows correspond to the 'from' trait
and columns correspond to the 'to' trait, so <code>G[1,2]</code> is the direct effect of trait 1 on trait 2. G should have 0
on the diagonal.</p></dd>


<dt>est_s</dt>
<dd><p>If TRUE, return estimates of se(`beta_hat`). If FALSE, the exact standard error of `beta_hat` is returned. Defaults to FALSE.</p></dd>


<dt>R_obs</dt>
<dd><p>Total observational correlation between traits. R_obs won't impact summary statistics unless there is sample overlap.
See Details for default behavior.</p></dd>


<dt>R_E</dt>
<dd><p>Total correlation of the environmental components of traits. Use only one of R_obs or R_E.
R_E may be phased out in the future.</p></dd>


<dt>R_LD</dt>
<dd><p>Optional list of LD blocks. R_LD should have class <code>list</code>.
Each element of R_LD can be either a) a matrix, b) a sparse matrix (class <code>dsCMatrix</code>) or c) an eigen decomposition (class <code>eigen</code>).
All elements should be correlation matrices, meaning that they have 1 on the diagonal and are positive definite. See Details and vignettes.</p></dd>


<dt>R_LD_eigen</dt>
<dd><p>Optionally, also supply list of eigen decompositions. The program will not check that this matches R_LD.</p></dd>


<dt>af</dt>
<dd><p>Optional vector of allele frequencies. If R_LD is not supplied, af can be a scalar, vector or function.
If af is a function it should take a single argument (n) and return a vector of n allele frequencies (See Examples).
If R_LD is supplied, af must be a vector with length equal to the size of the supplied LD pattern (See Examples).</p></dd>


<dt>snp_effect_function</dt>
<dd><p>Optional function to generate variant effects.
<code>snp_effect_function</code> can be a single function
or list of functions of length equal to the number of traits (see Details).</p></dd>


<dt>snp_info</dt>
<dd><p>Optional <code>data.frame</code> of variant information to be passed to variant effect functions. If <code>R_LD</code> is
specified, <code>snp_info</code> should have number of rows equal to the size of the supplied LD pattern. Otherwise <code>snp_info</code>
should have <code>J</code> rows.</p></dd>


<dt>sporadic_pleiotropy</dt>
<dd><p>Allow sporadic pleiotropy between traits. Defaults to TRUE.</p></dd>


<dt>pi_exact</dt>
<dd><p>If TRUE, the number of direct effect SNPs for each trait will be exactly equal to <code>round(pi*J)</code>.</p></dd>


<dt>h2_exact</dt>
<dd><p>If TRUE, the heritability of each trait will be exactly `h2`.</p></dd>


<dt>return_dat</dt>
<dd><p>Useful development option, not recommend for general users.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A list with the following elements:</p>


<p>Simulated effect estimates and standard errors are contained in matrices</p>


<p>+ <code>beta_hat</code>: Effect estimates for each trait</p>


<p>+ <code>se_beta_hat</code> Standard error of effect estimates, equal to sqrt(1/N_m *Var(G_j)).</p>


<p>+ <code>s_estimate</code> Estimate of <code>se_beta_hat</code>. Present only if <code>est_s = TRUE</code>.</p>


<p>Four matrices contain direct and total marginal and joint SNP-trait associations:
+ <code>direct_SNP_effects_marg</code> and <code>direct_SNP_effects_joint</code> give direct effects of SNPs on traits. These are the
same if there is no LD. If there is LD, <code>direct_SNP_effects_marg</code> is the direct component of the expected marginal association.</p>


<p>+ <code>beta_marg</code> and <code>beta_joint</code> give the total SNP effects (direct and indirect). These are the
same if there is no LD. If there is LD, <code>beta_marg</code> is the total expected marginal association. i.e. <code>beta_marg</code> is the
expected value of <code>beta_hat</code>.</p>


<p>Four matrices describe the covariance of traits and the row correlation of effect estimates</p>


<p>+ <code>Sigma_G</code> Genetic variance-covariance matrix. Diagonal elements are equal to heritability.</p>


<p>+ <code>Sigma_E</code> Environmental variance-covariance matrix.</p>


<p>+ <code>trait_corr</code> Population correlation of traits. <code>trait_corr = Sigma_G + Sigma_E</code>.</p>


<p>+ <code>R</code> Row correlation of <code>beta_hat - beta_marg</code>, equal to <code>trait_corr</code> scaled by the overlap proportion matrix.</p>


<p>Finally,</p>


<p>+ <code>snp_info</code> Is a data frame with variant information. If <code>R_LD</code> was omitted, <code>snp_info</code> contains only the allele frequency of
each variant. If <code>R_LD</code> was included, <code>snp_info</code> also contains block and replicate information as well as any information supplied to the
<code>snp_info</code> input parameter.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function generates GWAS summary statistics from a linear SEM specified by the matrix G.
The previous "xyz" mode is now deprecated. If you are used to using this function in xyz mode,
you can generate the corresponding G using the xyz_to_G function (see examples).</p>
<p>G should be square (#traits by #traits) matrix with 0s on the diagonal.
All traits have variance 1, so <code>G[i,j]^2</code> is the proportion of variance of trait j explained by the direct effect of trait i.
You will get an error if you specify a cyclic DAG. There are also some combinations of DAG and heritability that are impossible and
will give an error. For example in a two trait DAG where trait 1 has an effect of \(x\) on trait 2, and trait 1 has a heritability
of \(h^2_1\), trait 2 must have a heritability of at least \((x^2)*(h^2_1)\).</p>
<p>To generate data with LD, supply <code>R_LD</code> which can be a list of matrices, sparse matrices, or eigen-decompositions.
These matrices are interpreted as blocks in a block-diagonal LD matrix. The <code>af</code> argument must be provided
and should be a vector with length equal to the total size of the LD pattern (the sum of the sizes of each block).
<code>R_LD</code> does not need to have the same size as <code>J</code>. The LD pattern will be repeated or subset as necessary
to generate the desired number of variants.</p>
<p>If <code>R_obs</code> is NULL (default value), we assume that direct environmental effects on each trait are independent and all
environmental correlation results from the relationships specified in <code>G</code>. Alternatively <code>R_obs</code> can be any positive
definite correlation matrix.</p>
<p>The <code>snp_effect_function</code> argument supplies a custom function to generate standardized direct variant effects. By default,
standardized direct effects are generated from a normal distribution. This means that in the default mode, all variants
have equal expected heritability explained regardless of allele frequency or other features. The function(s) passed to
<code>snp_effect_function</code> will override this default. This function should take three arguments, <code>n</code>, <code>sd</code>, and <code>snp_info</code>.
The output should be a vector of length n with expected total sum of squares equal to <code>sd^2</code>. Note: Even though the
final effects returned will usually be on the per-allele scale, <code>snp_effect_function</code> should
return standardized (per-sd) effect sizes.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Two traits with no causal relationship and some environmental correlation</span></span></span>
<span class="r-in"><span><span class="co"># specify completely overlapping GWAS</span></span></span>
<span class="r-in"><span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1000</span>, nrow <span class="op">=</span> <span class="fl">2</span>, ncol <span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">R_obs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0.3</span>, <span class="fl">0.3</span>, <span class="fl">1</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">sim_mv</span><span class="op">(</span>G <span class="op">=</span> <span class="fl">2</span>, N <span class="op">=</span> <span class="va">N</span>, J <span class="op">=</span> <span class="fl">20000</span>, h2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.4</span>, <span class="fl">0.3</span><span class="op">)</span>, pi <span class="op">=</span> <span class="fl">1000</span><span class="op">/</span><span class="fl">20000</span>,</span></span>
<span class="r-in"><span>               R_obs <span class="op">=</span> <span class="va">R_obs</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> SNP effects provided for 20000 SNPs and 2 traits.</span>
<span class="r-in"><span><span class="va">dat</span><span class="op">$</span><span class="va">R</span> <span class="co"># This is the true correlation of the estimation error of beta_hat</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      [,1] [,2]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,]  1.0  0.3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,]  0.3  1.0</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">beta_hat</span> <span class="op">-</span> <span class="va">dat</span><span class="op">$</span><span class="va">beta_marg</span><span class="op">)</span> <span class="co"># Should be similar to dat$R</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           [,1]      [,2]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,] 1.0000000 0.2935818</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,] 0.2935818 1.0000000</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># The af argument can be a scalar, vector, or function.</span></span></span>
<span class="r-in"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">sim_mv</span><span class="op">(</span>N <span class="op">=</span> <span class="va">N</span>, J <span class="op">=</span> <span class="fl">20000</span>, h2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.4</span>, <span class="fl">0.3</span><span class="op">)</span>, pi <span class="op">=</span> <span class="fl">1000</span><span class="op">/</span><span class="fl">20000</span>,</span></span>
<span class="r-in"><span>               G <span class="op">=</span> <span class="fl">2</span>, R_obs <span class="op">=</span> <span class="va">R_obs</span>, af <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/Beta.html" class="external-link">rbeta</a></span><span class="op">(</span>n <span class="op">=</span> <span class="va">n</span>, <span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> SNP effects provided for 20000 SNPs and 2 traits.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># A very simple example with LD</span></span></span>
<span class="r-in"><span><span class="co"># Use a pattern of two small blocks of LD</span></span></span>
<span class="r-in"><span><span class="va">A1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0.7</span>, nrow <span class="op">=</span> <span class="fl">10</span>, ncol <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">A1</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span></span></span>
<span class="r-in"><span><span class="va">A2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0.1</span>, nrow <span class="op">=</span> <span class="fl">6</span>, ncol <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">A2</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span></span></span>
<span class="r-in"><span><span class="co"># If using LD, af should have the same size as the LD pattern</span></span></span>
<span class="r-in"><span><span class="va">af</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">16</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">sim_mv</span><span class="op">(</span>N <span class="op">=</span> <span class="va">N</span>, J <span class="op">=</span> <span class="fl">20000</span>, h2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.4</span>, <span class="fl">0.3</span><span class="op">)</span>, pi <span class="op">=</span> <span class="fl">1000</span><span class="op">/</span><span class="fl">20000</span>,</span></span>
<span class="r-in"><span>               G <span class="op">=</span> <span class="fl">2</span>, R_obs <span class="op">=</span> <span class="va">R_obs</span>, R_LD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">A1</span>, <span class="va">A2</span><span class="op">)</span>, af <span class="op">=</span> <span class="va">af</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> SNP effects provided for 20000 SNPs and 2 traits.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Use xyz_to_G to generate G from xyz specification</span></span></span>
<span class="r-in"><span><span class="va">myG</span> <span class="op">&lt;-</span> <span class="fu"><a href="xyz_to_G.html">xyz_to_G</a></span><span class="op">(</span>tau_xz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="op">-</span><span class="fl">0.3</span><span class="op">)</span>, tau_yz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.25</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>        dir_xz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="op">-</span><span class="fl">1</span><span class="op">)</span>, dir_yz <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>, gamma <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># If N is a scalar or a vector, there is no sample overlap</span></span></span>
<span class="r-in"><span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">sim_mv</span><span class="op">(</span>N <span class="op">=</span> <span class="fl">10000</span>, J <span class="op">=</span> <span class="fl">20000</span>, h2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0.4</span>, <span class="fl">4</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>              pi <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">500</span>, <span class="fl">500</span>, <span class="fl">1000</span>, <span class="fl">1000</span><span class="op">)</span><span class="op">/</span><span class="fl">20000</span>,</span></span>
<span class="r-in"><span>              G <span class="op">=</span> <span class="va">myG</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> SNP effects provided for 20000 SNPs and 4 traits.</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">dat</span><span class="op">$</span><span class="va">beta_marg</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span>, <span class="va">dat</span><span class="op">$</span><span class="va">beta_marg</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span><span class="fl">0</span>, <span class="va">dat</span><span class="op">$</span><span class="va">total_trait_effects</span><span class="op">[</span><span class="fl">3</span>,<span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="sim_mv-1.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jean Morrison.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

