<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Simulate multivariate GWAS Data with Specified Direct Effects"><title>Simulate multivariate GWAS Data with Specified Direct Effects — sim_mv_determined • GWASBrewer</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulate multivariate GWAS Data with Specified Direct Effects — sim_mv_determined"><meta property="og:description" content="Simulate multivariate GWAS Data with Specified Direct Effects"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">GWASBrewer</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.0.0237</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/effect_distribution.html">Controlling Effect Size Distributions</a>
    <a class="dropdown-item" href="../articles/resampling.html">Resampling and Re-Scaling Summary and Individual Level Data</a>
    <a class="dropdown-item" href="../articles/simulating_data.html">Simulating Data</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/jean997/GWASBrewer/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Simulate multivariate GWAS Data with Specified Direct Effects</h1>
      <small class="dont-index">Source: <a href="https://github.com/jean997/GWASBrewer/blob/HEAD/R/sim_mv_determined.R" class="external-link"><code>R/sim_mv_determined.R</code></a></small>
      <div class="d-none name"><code>sim_mv_determined.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Simulate multivariate GWAS Data with Specified Direct Effects</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">sim_mv_determined</span><span class="op">(</span></span>
<span>  <span class="va">N</span>,</span>
<span>  <span class="va">direct_SNP_effects_joint</span>,</span>
<span>  <span class="va">geno_scale</span>,</span>
<span>  <span class="va">pheno_sd</span>,</span>
<span>  G <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  est_s <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  R_obs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  R_E <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  R_LD <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  af <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>N</dt>
<dd><p>GWAS sample size. N can be a scalar, vector, or matrix. If N is a scalar, all GWAS have the same sample size
and there is no overlap between studies. If N is a vector, each element of N specifies the sample size of the corresponding
GWAS and there is no overlap between studies. If N is a matrix, N_ii specifies the sample size of study i
and N_ij specifies the number of samples present in both study i and study j. The elements of N must be positive
but non-integer values will not generate an error.</p></dd>


<dt>direct_SNP_effects_joint</dt>
<dd><p>Matrix of direct variant effects. Should be variants by traits.</p></dd>


<dt>geno_scale</dt>
<dd><p>Genotype scale of provided effects. Either "allele" or "sd".</p></dd>


<dt>pheno_sd</dt>
<dd><p>Phenotype standard deviation, a scalar or vector of length number of traits.</p></dd>


<dt>G</dt>
<dd><p>Matrix of direct effects. Rows correspond to the 'from' trait
and columns correspond to the 'to' trait, so <code>G[1,2]</code> is the direct effect of trait 1 on trait 2. G should have 0
on the diagonal. Be sure that <code>G</code> is on the same scale as the effect sizes.</p></dd>


<dt>est_s</dt>
<dd><p>If TRUE, return estimates of se(`beta_hat`). If FALSE, the exact standard error of `beta_hat` is returned. Defaults to FALSE.</p></dd>


<dt>R_obs</dt>
<dd><p>Total observational correlation between traits. R_obs won't impact summary statistics unless there is sample overlap.
See Details for default behavior.</p></dd>


<dt>R_E</dt>
<dd><p>Total correlation of the environmental components only. R_E and R_obs are alternative methods of specifying trait correlation.
Use only one of these two options.
R_E may be phased out in the future.</p></dd>


<dt>R_LD</dt>
<dd><p>Optional list of LD blocks. R_LD should have class <code>list</code>.
Each element of R_LD can be either a) a matrix, b) a sparse matrix (class <code>dsCMatrix</code>) or c) an eigen decomposition (class <code>eigen</code>).
All elements should be correlation matrices, meaning that they have 1 on the diagonal and are positive definite. See Details and vignettes.</p></dd>


<dt>af</dt>
<dd><p>Optional vector of allele frequencies. If R_LD is not supplied, af can be a scalar, vector or function.
If af is a function it should take a single argument (n) and return a vector of n allele frequencies (See Examples).
If R_LD is supplied, af must be a vector with length equal to the size of the supplied LD pattern (See Examples).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A <code>sim_mv</code> function. See <code><a href="sim_mv.html">?sim_mv</a></code> for details.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>A wrapper for <code>sim_mv</code>. See <code><a href="sim_mv.html">?sim_mv</a></code> and the "Providing an Exact Set of Direct Effects" section of the Effect Size vignette.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">G</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0.5</span>, <span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span>, byrow <span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">my_effects</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">0</span>, nrow <span class="op">=</span> <span class="fl">10</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">my_effects</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.008</span>, <span class="fl">0.01</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">my_effects</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">6</span>, <span class="fl">9</span><span class="op">)</span>, <span class="fl">2</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.02</span>, <span class="fl">0.06</span>, <span class="fl">0.009</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">my_effects</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         [,1]   [,2]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1,] -0.008  0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [2,]  0.000  0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [3,]  0.000 -0.020</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [4,]  0.000  0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [5,]  0.010  0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [6,]  0.000  0.060</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7,]  0.000  0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [8,]  0.000  0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [9,]  0.000  0.009</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [10,]  0.000  0.000</span>
<span class="r-in"><span><span class="co"># for fun, lets include some sample overlap</span></span></span>
<span class="r-in"><span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">40000</span>, <span class="fl">10000</span>, <span class="fl">10000</span>, <span class="fl">20000</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">sim_dat</span> <span class="op">&lt;-</span> <span class="fu">sim_mv_determined</span><span class="op">(</span>N <span class="op">=</span> <span class="va">N</span>,</span></span>
<span class="r-in"><span>                              direct_SNP_effects_joint <span class="op">=</span> <span class="va">my_effects</span>,</span></span>
<span class="r-in"><span>                              geno_scale <span class="op">=</span> <span class="st">"sd"</span>,</span></span>
<span class="r-in"><span>                              pheno_sd <span class="op">=</span> <span class="fl">1</span>,</span></span>
<span class="r-in"><span>                              G<span class="op">=</span><span class="va">G</span>,</span></span>
<span class="r-in"><span>                              est_s <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> SNP effects provided for 10 SNPs and 2 traits.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">sim_dat</span><span class="op">$</span><span class="va">direct_SNP_effects_joint</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         [,1]   [,2]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1,] -0.008  0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [2,]  0.000  0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [3,]  0.000 -0.020</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [4,]  0.000  0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [5,]  0.010  0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [6,]  0.000  0.060</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7,]  0.000  0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [8,]  0.000  0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [9,]  0.000  0.009</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [10,]  0.000  0.000</span>
<span class="r-in"><span><span class="va">sim_dat</span><span class="op">$</span><span class="va">beta_joint</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>         [,1]   [,2]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [1,] -0.008 -0.004</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [2,]  0.000  0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [3,]  0.000 -0.020</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [4,]  0.000  0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [5,]  0.010  0.005</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [6,]  0.000  0.060</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [7,]  0.000  0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [8,]  0.000  0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [9,]  0.000  0.009</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [10,]  0.000  0.000</span>
<span class="r-in"><span><span class="va">sim_dat</span><span class="op">$</span><span class="va">Sigma_G</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>          [,1]     [,2]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,] 0.000164 0.000082</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,] 0.000082 0.004122</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jean Morrison.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

